{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "af81fa37",
   "metadata": {
    "colab_type": "text",
    "id": "view-in-github"
   },
   "source": [
    "<a href=\"https://colab.research.google.com/github/sanjaynagi/AgamPrimer/blob/main/Primer-Design-in-Anopheles-gambiae.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "AM9EeHef6tAD",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "AM9EeHef6tAD",
    "outputId": "b33357a8-582f-422a-be52-3d72eb065d1c"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "^C\r\n",
      "\u001b[31mERROR: Operation cancelled by user\u001b[0m\r\n"
     ]
    }
   ],
   "source": [
    "# First, install some packages we require\n",
    "!pip install AgamPrimer==0.4.6 primer3-py malariagen_data kaleido gget -q "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "528ae85d",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "528ae85d",
    "outputId": "753faa12-9126-40e9-bbbb-245ed3ead12a"
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n",
       "  const css_urls = [];\n",
       "  \n",
       "\n",
       "  const inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n  const css_urls = [];\n  \n\n  const inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Import libraries \n",
    "import AgamPrimer\n",
    "import pandas as pd\n",
    "import malariagen_data\n",
    "import numpy as np\n",
    "import primer3\n",
    "import matplotlib.pyplot as plt\n",
    "import gget"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "47bde61a",
   "metadata": {
    "id": "47bde61a"
   },
   "source": [
    "##**AgamPrimer: Primer design considering genetic variation in *Anopheles gambiae***\n",
    "**Author**: [Sanjay Curtis Nagi](https://sanjaynagi.github.io/)    \n",
    "**Email**: sanjay.nagi@lstmed.ac.uk  \n",
    "\n",
    "Often, we would like to design primers for PCR applications, such as genotyping (PCR, TaqMan, LNA) or gene expression qPCR (SYBR, TaqMan, LNA). However, single nucleotide polymorphisms (SNPs) in primer binding sites can result in differences or failures in PCR amplification, referred to as null alleles. \n",
    "\n",
    "In general, mismatches caused by SNPs are more of a problem as you move towards the 3' end. I recommend reading a really good article on this topic on the IDT website - [Consider SNPs when designing PCR and qPCR assays](https://eu.idtdna.com/pages/education/decoded/article/considering-snps-when-designing-pcr-and-qpcr-assays). In *An. gambiae s.l*, there is a [huge amount of genetic variation](https://genome.cshlp.org/content/30/10/1533.full); a SNP found approximately every 1.9 bases (!), which makes considering SNPs even more important when designing molecular assays. Thanks to primer3-py and the fantastic malariagen_data package, we can do all of this in the cloud, hosted by google!  \n",
    "\n",
    "\\\n",
    "\n",
    "####**Google Colab**\n",
    "\n",
    "If you are unfamiliar with iPython notebooks and google colab, I encourage you to read the [FAQ](https://research.google.com/colaboratory/faq.html) and watch the following [introduction](https://www.youtube.com/watch?v=inN8seMm7UI). In general, the cells contain python code, and can be run by pressing the play button next to each cell, and should be run in order.\n",
    "\n",
    "You may want to save a copy of the notebook into your google drive (`file -> save a copy in Drive`)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "SV03ewwc9de8",
   "metadata": {
    "cellView": "form",
    "id": "SV03ewwc9de8"
   },
   "outputs": [],
   "source": [
    "#@title **Selecting Primer Parameters** { run: \"auto\" }\n",
    "#@markdown In the below cells, replace the values with those desired for your primers.\n",
    "\n",
    "assay_type = 'qPCR primers'           #@param [\"gDNA primers\", \"gDNA primers + probe\", \"probe\", \"qPCR primers\"]\n",
    "assay_name = 'coeae2f'        #@param {type:\"string\"}\n",
    "min_amplicon_size = 60        #@param {type:\"integer\"}\n",
    "max_amplicon_size = 120      #@param {type:\"integer\"}\n",
    "amplicon_size_range = [[min_amplicon_size, max_amplicon_size]]\n",
    "n_primer_pairs = 6            #@param {type:\"slider\", min:1, max:20, step:1}\n",
    "\n",
    "#@markdown    \n",
    "#@markdown target_loc is required for gDNA primers and probes, and transcript required for qPCR primers.\n",
    "contig =  \"2L\"                #@param ['2L', '2R', '3L', '3R', 'X']\n",
    "\n",
    "target_loc =  '2422652'                #@param {type:\"string\"}\n",
    "target_loc = int(target_loc)\n",
    "transcript =  'AGAP006228-RA'         #@param {type:\"string\"} \n",
    "\n",
    "\n",
    "if any(item in assay_type for item in ['gDNA', 'probe']):\n",
    "  assert target_loc > 0, \"Target location must be above 0 and less than the contig length\"\n",
    "elif assay_type == 'qPCR primers':\n",
    "  assert len(transcript) > 2, \"Transcript ID is not valid, should be vectorbase ID such as 'AGAP004707-RD'\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aa733775",
   "metadata": {
    "id": "aa733775"
   },
   "source": [
    "Load sequence data for chromosomal arm of choice, using the [malariagen_data API](https://malariagen.github.io/vector-data/ag3/api.html):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5fa0be8d",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "5fa0be8d",
    "outputId": "a5fbb80c-3cb9-4d6b-940e-fc79b52301e3"
   },
   "outputs": [
    {
     "data": {
      "application/javascript": [
       "\n",
       "(function(root) {\n",
       "  function now() {\n",
       "    return new Date();\n",
       "  }\n",
       "\n",
       "  const force = true;\n",
       "\n",
       "  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n",
       "    root._bokeh_onload_callbacks = [];\n",
       "    root._bokeh_is_loading = undefined;\n",
       "  }\n",
       "\n",
       "  const JS_MIME_TYPE = 'application/javascript';\n",
       "  const HTML_MIME_TYPE = 'text/html';\n",
       "  const EXEC_MIME_TYPE = 'application/vnd.bokehjs_exec.v0+json';\n",
       "  const CLASS_NAME = 'output_bokeh rendered_html';\n",
       "\n",
       "  /**\n",
       "   * Render data to the DOM node\n",
       "   */\n",
       "  function render(props, node) {\n",
       "    const script = document.createElement(\"script\");\n",
       "    node.appendChild(script);\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when an output is cleared or removed\n",
       "   */\n",
       "  function handleClearOutput(event, handle) {\n",
       "    const cell = handle.cell;\n",
       "\n",
       "    const id = cell.output_area._bokeh_element_id;\n",
       "    const server_id = cell.output_area._bokeh_server_id;\n",
       "    // Clean up Bokeh references\n",
       "    if (id != null && id in Bokeh.index) {\n",
       "      Bokeh.index[id].model.document.clear();\n",
       "      delete Bokeh.index[id];\n",
       "    }\n",
       "\n",
       "    if (server_id !== undefined) {\n",
       "      // Clean up Bokeh references\n",
       "      const cmd_clean = \"from bokeh.io.state import curstate; print(curstate().uuid_to_server['\" + server_id + \"'].get_sessions()[0].document.roots[0]._id)\";\n",
       "      cell.notebook.kernel.execute(cmd_clean, {\n",
       "        iopub: {\n",
       "          output: function(msg) {\n",
       "            const id = msg.content.text.trim();\n",
       "            if (id in Bokeh.index) {\n",
       "              Bokeh.index[id].model.document.clear();\n",
       "              delete Bokeh.index[id];\n",
       "            }\n",
       "          }\n",
       "        }\n",
       "      });\n",
       "      // Destroy server and session\n",
       "      const cmd_destroy = \"import bokeh.io.notebook as ion; ion.destroy_server('\" + server_id + \"')\";\n",
       "      cell.notebook.kernel.execute(cmd_destroy);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  /**\n",
       "   * Handle when a new output is added\n",
       "   */\n",
       "  function handleAddOutput(event, handle) {\n",
       "    const output_area = handle.output_area;\n",
       "    const output = handle.output;\n",
       "\n",
       "    // limit handleAddOutput to display_data with EXEC_MIME_TYPE content only\n",
       "    if ((output.output_type != \"display_data\") || (!Object.prototype.hasOwnProperty.call(output.data, EXEC_MIME_TYPE))) {\n",
       "      return\n",
       "    }\n",
       "\n",
       "    const toinsert = output_area.element.find(\".\" + CLASS_NAME.split(' ')[0]);\n",
       "\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"id\"] !== undefined) {\n",
       "      toinsert[toinsert.length - 1].firstChild.textContent = output.data[JS_MIME_TYPE];\n",
       "      // store reference to embed id on output_area\n",
       "      output_area._bokeh_element_id = output.metadata[EXEC_MIME_TYPE][\"id\"];\n",
       "    }\n",
       "    if (output.metadata[EXEC_MIME_TYPE][\"server_id\"] !== undefined) {\n",
       "      const bk_div = document.createElement(\"div\");\n",
       "      bk_div.innerHTML = output.data[HTML_MIME_TYPE];\n",
       "      const script_attrs = bk_div.children[0].attributes;\n",
       "      for (let i = 0; i < script_attrs.length; i++) {\n",
       "        toinsert[toinsert.length - 1].firstChild.setAttribute(script_attrs[i].name, script_attrs[i].value);\n",
       "        toinsert[toinsert.length - 1].firstChild.textContent = bk_div.children[0].textContent\n",
       "      }\n",
       "      // store reference to server id on output_area\n",
       "      output_area._bokeh_server_id = output.metadata[EXEC_MIME_TYPE][\"server_id\"];\n",
       "    }\n",
       "  }\n",
       "\n",
       "  function register_renderer(events, OutputArea) {\n",
       "\n",
       "    function append_mime(data, metadata, element) {\n",
       "      // create a DOM node to render to\n",
       "      const toinsert = this.create_output_subarea(\n",
       "        metadata,\n",
       "        CLASS_NAME,\n",
       "        EXEC_MIME_TYPE\n",
       "      );\n",
       "      this.keyboard_manager.register_events(toinsert);\n",
       "      // Render to node\n",
       "      const props = {data: data, metadata: metadata[EXEC_MIME_TYPE]};\n",
       "      render(props, toinsert[toinsert.length - 1]);\n",
       "      element.append(toinsert);\n",
       "      return toinsert\n",
       "    }\n",
       "\n",
       "    /* Handle when an output is cleared or removed */\n",
       "    events.on('clear_output.CodeCell', handleClearOutput);\n",
       "    events.on('delete.Cell', handleClearOutput);\n",
       "\n",
       "    /* Handle when a new output is added */\n",
       "    events.on('output_added.OutputArea', handleAddOutput);\n",
       "\n",
       "    /**\n",
       "     * Register the mime type and append_mime function with output_area\n",
       "     */\n",
       "    OutputArea.prototype.register_mime_type(EXEC_MIME_TYPE, append_mime, {\n",
       "      /* Is output safe? */\n",
       "      safe: true,\n",
       "      /* Index of renderer in `output_area.display_order` */\n",
       "      index: 0\n",
       "    });\n",
       "  }\n",
       "\n",
       "  // register the mime type if in Jupyter Notebook environment and previously unregistered\n",
       "  if (root.Jupyter !== undefined) {\n",
       "    const events = require('base/js/events');\n",
       "    const OutputArea = require('notebook/js/outputarea').OutputArea;\n",
       "\n",
       "    if (OutputArea.prototype.mime_types().indexOf(EXEC_MIME_TYPE) == -1) {\n",
       "      register_renderer(events, OutputArea);\n",
       "    }\n",
       "  }\n",
       "\n",
       "  \n",
       "  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n",
       "    root._bokeh_timeout = Date.now() + 5000;\n",
       "    root._bokeh_failed_load = false;\n",
       "  }\n",
       "\n",
       "  const NB_LOAD_WARNING = {'data': {'text/html':\n",
       "     \"<div style='background-color: #fdd'>\\n\"+\n",
       "     \"<p>\\n\"+\n",
       "     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n",
       "     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n",
       "     \"</p>\\n\"+\n",
       "     \"<ul>\\n\"+\n",
       "     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n",
       "     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n",
       "     \"</ul>\\n\"+\n",
       "     \"<code>\\n\"+\n",
       "     \"from bokeh.resources import INLINE\\n\"+\n",
       "     \"output_notebook(resources=INLINE)\\n\"+\n",
       "     \"</code>\\n\"+\n",
       "     \"</div>\"}};\n",
       "\n",
       "  function display_loaded() {\n",
       "    const el = document.getElementById(null);\n",
       "    if (el != null) {\n",
       "      el.textContent = \"BokehJS is loading...\";\n",
       "    }\n",
       "    if (root.Bokeh !== undefined) {\n",
       "      if (el != null) {\n",
       "        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n",
       "      }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(display_loaded, 100)\n",
       "    }\n",
       "  }\n",
       "\n",
       "\n",
       "  function run_callbacks() {\n",
       "    try {\n",
       "      root._bokeh_onload_callbacks.forEach(function(callback) {\n",
       "        if (callback != null)\n",
       "          callback();\n",
       "      });\n",
       "    } finally {\n",
       "      delete root._bokeh_onload_callbacks\n",
       "    }\n",
       "    console.debug(\"Bokeh: all callbacks have finished\");\n",
       "  }\n",
       "\n",
       "  function load_libs(css_urls, js_urls, callback) {\n",
       "    if (css_urls == null) css_urls = [];\n",
       "    if (js_urls == null) js_urls = [];\n",
       "\n",
       "    root._bokeh_onload_callbacks.push(callback);\n",
       "    if (root._bokeh_is_loading > 0) {\n",
       "      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n",
       "      return null;\n",
       "    }\n",
       "    if (js_urls == null || js_urls.length === 0) {\n",
       "      run_callbacks();\n",
       "      return null;\n",
       "    }\n",
       "    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n",
       "    root._bokeh_is_loading = css_urls.length + js_urls.length;\n",
       "\n",
       "    function on_load() {\n",
       "      root._bokeh_is_loading--;\n",
       "      if (root._bokeh_is_loading === 0) {\n",
       "        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n",
       "        run_callbacks()\n",
       "      }\n",
       "    }\n",
       "\n",
       "    function on_error(url) {\n",
       "      console.error(\"failed to load \" + url);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < css_urls.length; i++) {\n",
       "      const url = css_urls[i];\n",
       "      const element = document.createElement(\"link\");\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.rel = \"stylesheet\";\n",
       "      element.type = \"text/css\";\n",
       "      element.href = url;\n",
       "      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n",
       "      document.body.appendChild(element);\n",
       "    }\n",
       "\n",
       "    for (let i = 0; i < js_urls.length; i++) {\n",
       "      const url = js_urls[i];\n",
       "      const element = document.createElement('script');\n",
       "      element.onload = on_load;\n",
       "      element.onerror = on_error.bind(null, url);\n",
       "      element.async = false;\n",
       "      element.src = url;\n",
       "      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n",
       "      document.head.appendChild(element);\n",
       "    }\n",
       "  };\n",
       "\n",
       "  function inject_raw_css(css) {\n",
       "    const element = document.createElement(\"style\");\n",
       "    element.appendChild(document.createTextNode(css));\n",
       "    document.body.appendChild(element);\n",
       "  }\n",
       "\n",
       "  \n",
       "  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n",
       "  const css_urls = [];\n",
       "  \n",
       "\n",
       "  const inline_js = [\n",
       "    function(Bokeh) {\n",
       "      Bokeh.set_log_level(\"info\");\n",
       "    },\n",
       "    function(Bokeh) {\n",
       "    \n",
       "    \n",
       "    }\n",
       "  ];\n",
       "\n",
       "  function run_inline_js() {\n",
       "    \n",
       "    if (root.Bokeh !== undefined || force === true) {\n",
       "      \n",
       "    for (let i = 0; i < inline_js.length; i++) {\n",
       "      inline_js[i].call(root, root.Bokeh);\n",
       "    }\n",
       "    } else if (Date.now() < root._bokeh_timeout) {\n",
       "      setTimeout(run_inline_js, 100);\n",
       "    } else if (!root._bokeh_failed_load) {\n",
       "      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n",
       "      root._bokeh_failed_load = true;\n",
       "    } else if (force !== true) {\n",
       "      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n",
       "      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n",
       "    }\n",
       "\n",
       "  }\n",
       "\n",
       "  if (root._bokeh_is_loading === 0) {\n",
       "    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n",
       "    run_inline_js();\n",
       "  } else {\n",
       "    load_libs(css_urls, js_urls, function() {\n",
       "      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n",
       "      run_inline_js();\n",
       "    });\n",
       "  }\n",
       "}(window));"
      ],
      "application/vnd.bokehjs_load.v0+json": "\n(function(root) {\n  function now() {\n    return new Date();\n  }\n\n  const force = true;\n\n  if (typeof root._bokeh_onload_callbacks === \"undefined\" || force === true) {\n    root._bokeh_onload_callbacks = [];\n    root._bokeh_is_loading = undefined;\n  }\n\n  \n\n  \n  if (typeof (root._bokeh_timeout) === \"undefined\" || force === true) {\n    root._bokeh_timeout = Date.now() + 5000;\n    root._bokeh_failed_load = false;\n  }\n\n  const NB_LOAD_WARNING = {'data': {'text/html':\n     \"<div style='background-color: #fdd'>\\n\"+\n     \"<p>\\n\"+\n     \"BokehJS does not appear to have successfully loaded. If loading BokehJS from CDN, this \\n\"+\n     \"may be due to a slow or bad network connection. Possible fixes:\\n\"+\n     \"</p>\\n\"+\n     \"<ul>\\n\"+\n     \"<li>re-rerun `output_notebook()` to attempt to load from CDN again, or</li>\\n\"+\n     \"<li>use INLINE resources instead, as so:</li>\\n\"+\n     \"</ul>\\n\"+\n     \"<code>\\n\"+\n     \"from bokeh.resources import INLINE\\n\"+\n     \"output_notebook(resources=INLINE)\\n\"+\n     \"</code>\\n\"+\n     \"</div>\"}};\n\n  function display_loaded() {\n    const el = document.getElementById(null);\n    if (el != null) {\n      el.textContent = \"BokehJS is loading...\";\n    }\n    if (root.Bokeh !== undefined) {\n      if (el != null) {\n        el.textContent = \"BokehJS \" + root.Bokeh.version + \" successfully loaded.\";\n      }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(display_loaded, 100)\n    }\n  }\n\n\n  function run_callbacks() {\n    try {\n      root._bokeh_onload_callbacks.forEach(function(callback) {\n        if (callback != null)\n          callback();\n      });\n    } finally {\n      delete root._bokeh_onload_callbacks\n    }\n    console.debug(\"Bokeh: all callbacks have finished\");\n  }\n\n  function load_libs(css_urls, js_urls, callback) {\n    if (css_urls == null) css_urls = [];\n    if (js_urls == null) js_urls = [];\n\n    root._bokeh_onload_callbacks.push(callback);\n    if (root._bokeh_is_loading > 0) {\n      console.debug(\"Bokeh: BokehJS is being loaded, scheduling callback at\", now());\n      return null;\n    }\n    if (js_urls == null || js_urls.length === 0) {\n      run_callbacks();\n      return null;\n    }\n    console.debug(\"Bokeh: BokehJS not loaded, scheduling load and callback at\", now());\n    root._bokeh_is_loading = css_urls.length + js_urls.length;\n\n    function on_load() {\n      root._bokeh_is_loading--;\n      if (root._bokeh_is_loading === 0) {\n        console.debug(\"Bokeh: all BokehJS libraries/stylesheets loaded\");\n        run_callbacks()\n      }\n    }\n\n    function on_error(url) {\n      console.error(\"failed to load \" + url);\n    }\n\n    for (let i = 0; i < css_urls.length; i++) {\n      const url = css_urls[i];\n      const element = document.createElement(\"link\");\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.rel = \"stylesheet\";\n      element.type = \"text/css\";\n      element.href = url;\n      console.debug(\"Bokeh: injecting link tag for BokehJS stylesheet: \", url);\n      document.body.appendChild(element);\n    }\n\n    for (let i = 0; i < js_urls.length; i++) {\n      const url = js_urls[i];\n      const element = document.createElement('script');\n      element.onload = on_load;\n      element.onerror = on_error.bind(null, url);\n      element.async = false;\n      element.src = url;\n      console.debug(\"Bokeh: injecting script tag for BokehJS library: \", url);\n      document.head.appendChild(element);\n    }\n  };\n\n  function inject_raw_css(css) {\n    const element = document.createElement(\"style\");\n    element.appendChild(document.createTextNode(css));\n    document.body.appendChild(element);\n  }\n\n  \n  const js_urls = [\"https://cdn.bokeh.org/bokeh/release/bokeh-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-gl-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-widgets-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-tables-2.4.2.min.js\", \"https://cdn.bokeh.org/bokeh/release/bokeh-mathjax-2.4.2.min.js\"];\n  const css_urls = [];\n  \n\n  const inline_js = [\n    function(Bokeh) {\n      Bokeh.set_log_level(\"info\");\n    },\n    function(Bokeh) {\n    \n    \n    }\n  ];\n\n  function run_inline_js() {\n    \n    if (root.Bokeh !== undefined || force === true) {\n      \n    for (let i = 0; i < inline_js.length; i++) {\n      inline_js[i].call(root, root.Bokeh);\n    }\n    } else if (Date.now() < root._bokeh_timeout) {\n      setTimeout(run_inline_js, 100);\n    } else if (!root._bokeh_failed_load) {\n      console.log(\"Bokeh: BokehJS failed to load within specified timeout.\");\n      root._bokeh_failed_load = true;\n    } else if (force !== true) {\n      const cell = $(document.getElementById(null)).parents('.cell').data().cell;\n      cell.output_area.append_execute_result(NB_LOAD_WARNING)\n    }\n\n  }\n\n  if (root._bokeh_is_loading === 0) {\n    console.debug(\"Bokeh: BokehJS loaded, going straight to plotting\");\n    run_inline_js();\n  } else {\n    load_libs(css_urls, js_urls, function() {\n      console.debug(\"Bokeh: BokehJS plotting callback run at\", now());\n      run_inline_js();\n    });\n  }\n}(window));"
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Our genome sequence for 2L is 49364325 bp long\n"
     ]
    }
   ],
   "source": [
    "# Connect to the malariagen_data ag3 API\n",
    "ag3 = malariagen_data.Ag3() #pre=True\n",
    "genome_seq = ag3.genome_sequence(region=contig)\n",
    "print(f\"Our genome sequence for {contig} is {genome_seq.shape[0]} bp long\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "144a37ce",
   "metadata": {
    "id": "144a37ce"
   },
   "source": [
    "Now we need to extract the bit of sequence we need. We will use functions in the [AgamPrimer](https://pypi.org/project/AgamPrimer/) package."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "R0CkEd38VXGY",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "R0CkEd38VXGY",
    "outputId": "5f8509f0-a2c3-42b4-cd92-10e3ee97d016"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue Aug 30 21:27:39 2022 DEBUG geneset: handle region\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Exon junctions for AGAP006228-RA: [ 330  467 1009] [28548763, 28549199, 28549818] \n",
      "\n"
     ]
    }
   ],
   "source": [
    "if any(item in assay_type for item in ['gDNA', 'probe']):\n",
    "  # genomic DNA\n",
    "  target_sequence, gdna_pos, seq_parameters = AgamPrimer.prepare_gDNA_sequence(target_loc=target_loc, amplicon_size_range=amplicon_size_range, genome_seq=genome_seq, assay_name=assay_name, assay_type=assay_type)\n",
    "elif assay_type == 'qPCR primers':\n",
    "  # quantitative PCR \n",
    "  target_sequence, exon_junctions, gdna_pos, seq_parameters = AgamPrimer.prepare_cDNA_sequence(transcript=transcript, gff=ag3.geneset(), genome_seq=genome_seq, assay_name=assay_name)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "_aaCJGAXUCSx",
   "metadata": {
    "id": "_aaCJGAXUCSx"
   },
   "source": [
    "Now we have our target sequence. Lets take a look..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "nDaWXq7f9hyA",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 113
    },
    "id": "nDaWXq7f9hyA",
    "outputId": "87738985-33ee-4dc2-ba52-195d4ac095b9"
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CCGCGGCCTACTGCACAACCACAACTCACGAACCACTTGTCATCTAGAAATCTTCCAACAGAGGTGGTTCAGTTTACTTCTTGCACGGTTCGGGTTGTGATTGTATCAGAGCAGTTGGTGCTGTGTTGTGTCCCTTTTTTGTTCGTCTGATATTTTAGAGCTTTGAAAATTCAAATCATATCAAGCTGCAATGGCCAACAGCGAGCTGATTGTTAGCACCGGCTACGGACCGGTTCAGGGCACAGCCAGGACATCGCTCTACGGGACTGGGTACGTCAGCTTCCAGGGCATCCCGTACGCGAAACCTCCGGTCGGAGAGCTACGTTTTAAGGATCCCACGCCACCTGAAAACTGGACGCAGGTTCTCGATTGTACCGAGCAATGTGACCCTTGCTTTCATTTTGATCGGCGCGTTAACAAAATTGTCGGGAGTGAGGACAGCCTGAGGCTGAACATATTCAGCAAAACGATTAAACCCACCAAACCTCTGCCCGTCATGGTGTACATCTACGGCGGTGGATTCGTCGAAGGTACCAGCGGAACCGAACTGTACGGTCCGGACTATCTGATCGAAAAAGACATCGTGCTGGTAACGCTTAACTATCGCGTCGGAGCACTCGGTTTCCTCTGCTGCCAATCACCCACGGCAGGTGTCCCAGGCAATGCGGGCCTGAAAGACCAAAGGCTGGCGCTCCGATGGGTGCGGGACAACATTGCGTCCTTCGGAGGCGATCCCAGCGCGATCACCCTGTTTGGACACAGCGCCGGGGGAGCGTCGGTACAGTATCACACGATTGCGGACGCGTCGAAGAATTTGTTCCAGCGTGCCATCATCATGTCCGGCAGCACGATGTGCAGCTGGGCCCTTACACCCCAGCGCAACTGGCCGGAAAAACTGGCCAAAGCCATCGGTTGGCAGGGCGAGGGTGATGAGGAGGCGGCGCTGCAATATTTGCGCCAAGCTTCGCCGGAAAGCATTGTGGATCATCAGGAGAAATTGTTCGGACCGCAGGAAATACAAGAAGGCCTGCTTTCTCCATTCGCGCCAACGATCGAACCGTACGAGAGTGAGGTATGTTTCATCCCACGATCGCCGTTCGAGATGTCCCGCACTGCGTGGGGAAATTCCATCGACATCATGATCGGCGGAACCTCCGAAGAGGGGCTGATACTGCTGCCGAAAGTGAAACCACAGCTTCCAAGCATGCTGCAGGATCCGCGACTGTTTGTAGGCAATGTTCCCTTCCATCTCAAGCTTAGCCTAGAGCAGCGTATGGCTTTCGGGGAGCAGCTGAAGCAACTCTACTATCCAGATTCCAACCCCTCGATCGACAACCTGGATGGATTCGTGAACATGGCCTCGGATCGCATCTTTTGGCACGATCTGCATCGTACCATACTGGCACGGGCCAACTACGCCTGCACTGCCAAAACGTTCGTGTATCGCTTCTGCGTTGATTCTCCGTTCTTTAACCACTATCGTATCCACATGGTCGACCCGAACGCACGTGGGACGTCTCATGCGGATGAAATTTCTTACCTGTTTTCAAACATCTTCGCGAAACCGCTGGACAAGAGCACGCTCGAGTACCGTGCCATCCAGCACTTGGTCGATATCTTCACCAGTTTTGCCACCAACAGCGATCCCAACTGTGATAGCACCGCCAGTCTGTCGTGGACGGCCGTCCCGAAAACGGCACCGCCGTACAATTGCCTGAACATCTCCAACGATGGTGTGGAGGTTGTGGAGCTGCCGGAATCAAGACGTCTGCAGCTGTGGGATTCGTTCTACGTAAATGATGCACTCTTCTAGACAGTACGATATTATTCGACGAAATTTTGAACAACAATAAATTCTATATGAATCGTGTGC'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "target_sequence"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "38fe2530",
   "metadata": {
    "id": "38fe2530"
   },
   "source": [
    "We need to set up some python dictionaries containing our sequence and primer parameters, this will be our input to primer3. In the below cell, you can modify or add primer3 parameters, such as optimal TM, GC content etc etc. A full list of possible parameters and their functions can be found in the [primer3 manual](https://primer3.org/manual.html)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "622ac6d9",
   "metadata": {
    "id": "622ac6d9"
   },
   "outputs": [],
   "source": [
    "primer_parameters  =  {\n",
    "        'PRIMER_NUM_RETURN':n_primer_pairs,\n",
    "        'PRIMER_OPT_SIZE': 20,\n",
    "        'PRIMER_TASK':'generic',\n",
    "        'PRIMER_MIN_SIZE': 17,\n",
    "        'PRIMER_MAX_SIZE': 24,\n",
    "        'PRIMER_OPT_TM': 60.0,\n",
    "        'PRIMER_MIN_TM': 57.0,\n",
    "        'PRIMER_MAX_TM': 63.0,\n",
    "        'PRIMER_MIN_GC': 30.0,\n",
    "        'PRIMER_MAX_GC': 75.0,\n",
    "        'PRIMER_PRODUCT_SIZE_RANGE': amplicon_size_range,\n",
    "        'PRIMER_MIN_THREE_PRIME_DISTANCE':3,          # this parameter is the minimum distance between successive pairs. If 1, it means successive primer pairs could be identical bar one base shift\n",
    "        'PRIMER_INTERNAL_OPT_SIZE': 16,               # Probe size preferences if selected, otherwise ignored\n",
    "        'PRIMER_INTERNAL_MIN_SIZE': 10,\n",
    "        'PRIMER_INTERNAL_MAX_SIZE': 22,\n",
    "        'PRIMER_INTERNAL_MIN_TM': 45,\n",
    "        'PRIMER_INTERNAL_MAX_TM':65,                # Probe considerations are quite relaxed, assumed that LNAs will be used later to affect TM\n",
    "        # Extra primer3 parameters can go here\n",
    "        # In the same format as above                       \n",
    "    }\n",
    "\n",
    "primer_parameters = AgamPrimer.primer_params(primer_parameters, assay_type) ## adds some parameters depending on assay type"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0819ba76",
   "metadata": {
    "id": "0819ba76"
   },
   "source": [
    "#### **Run the primer3 algorithm!**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ae6e7525",
   "metadata": {
    "id": "ae6e7525",
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "primer_dict = primer3.designPrimers(seq_args=seq_parameters, global_args=primer_parameters)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56945ca6",
   "metadata": {
    "id": "56945ca6"
   },
   "source": [
    "It should be *fast*. The output, which we call 'primer_dict', is a python dictionary containing the full results from primer3. We will turn this into a pandas dataframe (i.e a useful, pretty table), containing just the necessary bits of information. First, we'll print some information from the primer3 run."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1bb40846",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "1bb40846",
    "outputId": "612e982a-b944-4c12-e489-00770b65e7f1"
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PRIMER_LEFT_EXPLAIN  :  considered 14417, GC content failed 492, low tm 3723, high tm 4772, high hairpin stability 61, long poly-x seq 16, ok 5353 \n",
      "\n",
      "PRIMER_RIGHT_EXPLAIN  :  considered 14414, GC content failed 724, low tm 3643, high tm 4729, long poly-x seq 18, ok 5300 \n",
      "\n",
      "PRIMER_PAIR_EXPLAIN  :  considered 471002, unacceptable product size 5806, no overlap of required point 465184, primer in pair overlaps a primer in a better pair 7687, ok 12 \n",
      "\n",
      "PRIMER_LEFT_NUM_RETURNED  :  6 \n",
      "\n",
      "PRIMER_RIGHT_NUM_RETURNED  :  6 \n",
      "\n",
      "PRIMER_INTERNAL_NUM_RETURNED  :  0 \n",
      "\n",
      "PRIMER_PAIR_NUM_RETURNED  :  6 \n",
      "\n"
     ]
    }
   ],
   "source": [
    "AgamPrimer.primer3_run_statistics(primer_dict, assay_type)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "CbVMwOdCrNMf",
   "metadata": {
    "id": "CbVMwOdCrNMf"
   },
   "source": [
    "Now lets wrangle this into an easy to look at table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "bf8c5970",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 328
    },
    "id": "bf8c5970",
    "outputId": "cf0dc28e-f37c-412d-e9f1-8ecdc93a18c9"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>primer_pair</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>parameter</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PRIMER_LEFT_SEQUENCE</th>\n",
       "      <td>GACATCGCTCTACGGGACTG</td>\n",
       "      <td>CGTTTTAAGGATCCCACGCC</td>\n",
       "      <td>CTGGGTACGTCAGCTTCCAG</td>\n",
       "      <td>CTTCGCCGGAAAGCATTGTG</td>\n",
       "      <td>GCCGGAAAGCATTGTGGATC</td>\n",
       "      <td>TTTTAAGGATCCCACGCCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_RIGHT_SEQUENCE</th>\n",
       "      <td>GGTGGCGTGGGATCCTTAAA</td>\n",
       "      <td>GGGTCACATTGCTCGGTACA</td>\n",
       "      <td>GGCGTGGGATCCTTAAAACG</td>\n",
       "      <td>CTTGTATTTCCTGCGGTCCG</td>\n",
       "      <td>GCCTTCTTGTATTTCCTGCGG</td>\n",
       "      <td>ACGCGCCGATCAAAATGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_LEFT_TM</th>\n",
       "      <td>59.970433</td>\n",
       "      <td>59.26916</td>\n",
       "      <td>60.108615</td>\n",
       "      <td>60.454957</td>\n",
       "      <td>59.898781</td>\n",
       "      <td>60.61485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_RIGHT_TM</th>\n",
       "      <td>60.034863</td>\n",
       "      <td>60.036765</td>\n",
       "      <td>59.26916</td>\n",
       "      <td>58.993168</td>\n",
       "      <td>59.598095</td>\n",
       "      <td>59.480481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_LEFT_GC_PERCENT</th>\n",
       "      <td>60.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>52.380952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_RIGHT_GC_PERCENT</th>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>52.380952</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_LEFT</th>\n",
       "      <td>(249, 20)</td>\n",
       "      <td>(322, 20)</td>\n",
       "      <td>(266, 20)</td>\n",
       "      <td>(962, 20)</td>\n",
       "      <td>(966, 20)</td>\n",
       "      <td>(324, 21)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_RIGHT</th>\n",
       "      <td>(344, 20)</td>\n",
       "      <td>(389, 20)</td>\n",
       "      <td>(341, 20)</td>\n",
       "      <td>(1021, 20)</td>\n",
       "      <td>(1026, 21)</td>\n",
       "      <td>(413, 20)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_PAIR_PRODUCT_SIZE</th>\n",
       "      <td>96</td>\n",
       "      <td>68</td>\n",
       "      <td>76</td>\n",
       "      <td>60</td>\n",
       "      <td>61</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "primer_pair                                  0                     1  \\\n",
       "parameter                                                              \n",
       "PRIMER_LEFT_SEQUENCE      GACATCGCTCTACGGGACTG  CGTTTTAAGGATCCCACGCC   \n",
       "PRIMER_RIGHT_SEQUENCE     GGTGGCGTGGGATCCTTAAA  GGGTCACATTGCTCGGTACA   \n",
       "PRIMER_LEFT_TM                       59.970433              59.26916   \n",
       "PRIMER_RIGHT_TM                      60.034863             60.036765   \n",
       "PRIMER_LEFT_GC_PERCENT                    60.0                  55.0   \n",
       "PRIMER_RIGHT_GC_PERCENT                   55.0                  55.0   \n",
       "PRIMER_LEFT                          (249, 20)             (322, 20)   \n",
       "PRIMER_RIGHT                         (344, 20)             (389, 20)   \n",
       "PRIMER_PAIR_PRODUCT_SIZE                    96                    68   \n",
       "\n",
       "primer_pair                                  2                     3  \\\n",
       "parameter                                                              \n",
       "PRIMER_LEFT_SEQUENCE      CTGGGTACGTCAGCTTCCAG  CTTCGCCGGAAAGCATTGTG   \n",
       "PRIMER_RIGHT_SEQUENCE     GGCGTGGGATCCTTAAAACG  CTTGTATTTCCTGCGGTCCG   \n",
       "PRIMER_LEFT_TM                       60.108615             60.454957   \n",
       "PRIMER_RIGHT_TM                       59.26916             58.993168   \n",
       "PRIMER_LEFT_GC_PERCENT                    60.0                  55.0   \n",
       "PRIMER_RIGHT_GC_PERCENT                   55.0                  55.0   \n",
       "PRIMER_LEFT                          (266, 20)             (962, 20)   \n",
       "PRIMER_RIGHT                         (341, 20)            (1021, 20)   \n",
       "PRIMER_PAIR_PRODUCT_SIZE                    76                    60   \n",
       "\n",
       "primer_pair                                   4                      5  \n",
       "parameter                                                               \n",
       "PRIMER_LEFT_SEQUENCE       GCCGGAAAGCATTGTGGATC  TTTTAAGGATCCCACGCCACC  \n",
       "PRIMER_RIGHT_SEQUENCE     GCCTTCTTGTATTTCCTGCGG   ACGCGCCGATCAAAATGAAA  \n",
       "PRIMER_LEFT_TM                        59.898781               60.61485  \n",
       "PRIMER_RIGHT_TM                       59.598095              59.480481  \n",
       "PRIMER_LEFT_GC_PERCENT                     55.0              52.380952  \n",
       "PRIMER_RIGHT_GC_PERCENT               52.380952                   45.0  \n",
       "PRIMER_LEFT                           (966, 20)              (324, 21)  \n",
       "PRIMER_RIGHT                         (1026, 21)              (413, 20)  \n",
       "PRIMER_PAIR_PRODUCT_SIZE                     61                     90  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primer_df = AgamPrimer.primer3_to_pandas(primer_dict, assay_type)\n",
    "primer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "Lf0CeEVrtaxP",
   "metadata": {
    "id": "Lf0CeEVrtaxP"
   },
   "source": [
    "\n",
    "We can write this to .tsv and excel files, which can be explored in other editors. To download a file from colab to your local computer, click the folder panel on the left-hand sidebar, the three dots next your primers.tsv/.xlsx file, and download."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "09eQYSeT7lHG",
   "metadata": {
    "id": "09eQYSeT7lHG"
   },
   "outputs": [],
   "source": [
    "primer_df.to_csv(f\"{assay_name}.{assay_type}.primers.tsv\", sep=\"\\t\")\n",
    "primer_df.to_excel(f\"{assay_name}.{assay_type}.primers.xlsx\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b4a9776c",
   "metadata": {
    "id": "b4a9776c"
   },
   "source": [
    "##**Looking for variation using the ag1000g resource and malariagen API**\n",
    "\n",
    "In Ag3, samples are organised into sample sets. We can load any sample set from the Ag3 resource, but there are quite a few! Lets look at what each sample set contains, breaking it down by species, year and country. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "61286685",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 49,
     "referenced_widgets": [
      "e2fec3b726cd4f529b58b0716f2237a3",
      "506dbb25d1f743ee80033d6f98dabc3b",
      "5a98a29b6ecf41f09fee8d7aebf3132e",
      "743f1f0ae54f4f4aa4994c96a523625c",
      "792f088bb26346baa2f17f4cfcafe0ea",
      "2db4d9dd91a24e32b80d870e9869339a",
      "cbf3c833b87d4f51b10aa6f09d53d5d1",
      "a60370f9435b42ab8b5d794ce32c2fa2",
      "cc3c6714c3ab435ca820f8081878dc3f",
      "4546f79977e24400806287bebf4efca0",
      "b913e8c4b6ca4c129949878e1b832774"
     ]
    },
    "id": "61286685",
    "outputId": "406f20a5-93b8-4c37-d051-a14b6e8a5ace"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "e2fec3b726cd4f529b58b0716f2237a3",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "Load sample metadata:   0%|          | 0/28 [00:00<?, ?it/s]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "metadata = ag3.sample_metadata()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "iSTRbVpxD3oy",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "iSTRbVpxD3oy",
    "outputId": "86a6b0ee-3dc9-4306-9f8b-d9deffac3b65"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-ea66cd18-8f81-4b50-ac93-c2f7e71b4dc9\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>taxon</th>\n",
       "      <th>arabiensis</th>\n",
       "      <th>coluzzii</th>\n",
       "      <th>gambiae</th>\n",
       "      <th>gcx1</th>\n",
       "      <th>gcx2</th>\n",
       "      <th>gcx3</th>\n",
       "      <th>intermediate_gambcolu_arabiensis</th>\n",
       "      <th>intermediate_gambiae_coluzzii</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_set</th>\n",
       "      <th>year</th>\n",
       "      <th>country</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AG1000G-AO</th>\n",
       "      <th>2009</th>\n",
       "      <th>Angola</th>\n",
       "      <td>0</td>\n",
       "      <td>81</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-BF-A</th>\n",
       "      <th>2012</th>\n",
       "      <th>Burkina Faso</th>\n",
       "      <td>0</td>\n",
       "      <td>82</td>\n",
       "      <td>98</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-BF-B</th>\n",
       "      <th>2014</th>\n",
       "      <th>Burkina Faso</th>\n",
       "      <td>3</td>\n",
       "      <td>53</td>\n",
       "      <td>46</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-BF-C</th>\n",
       "      <th>2004</th>\n",
       "      <th>Burkina Faso</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-CD</th>\n",
       "      <th>2015</th>\n",
       "      <th>Democratic Republic of the Congo</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>76</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">AG1000G-CF</th>\n",
       "      <th>1993</th>\n",
       "      <th>Central African Republic</th>\n",
       "      <td>0</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1994</th>\n",
       "      <th>Central African Republic</th>\n",
       "      <td>0</td>\n",
       "      <td>13</td>\n",
       "      <td>53</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-CI</th>\n",
       "      <th>2012</th>\n",
       "      <th>Cote d'Ivoire</th>\n",
       "      <td>0</td>\n",
       "      <td>80</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-CM-A</th>\n",
       "      <th>2009</th>\n",
       "      <th>Cameroon</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>303</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-CM-B</th>\n",
       "      <th>2005</th>\n",
       "      <th>Cameroon</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>90</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-CM-C</th>\n",
       "      <th>2013</th>\n",
       "      <th>Cameroon</th>\n",
       "      <td>2</td>\n",
       "      <td>19</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-FR</th>\n",
       "      <th>2011</th>\n",
       "      <th>Mayotte</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-GA-A</th>\n",
       "      <th>2000</th>\n",
       "      <th>Gabon</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>69</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-GH</th>\n",
       "      <th>2012</th>\n",
       "      <th>Ghana</th>\n",
       "      <td>0</td>\n",
       "      <td>64</td>\n",
       "      <td>36</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-GM-A</th>\n",
       "      <th>2011</th>\n",
       "      <th>Gambia, The</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>68</td>\n",
       "      <td>6</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-GM-B</th>\n",
       "      <th>2012</th>\n",
       "      <th>Gambia, The</th>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>9</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-GM-C</th>\n",
       "      <th>2012</th>\n",
       "      <th>Gambia, The</th>\n",
       "      <td>0</td>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>160</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-GN-A</th>\n",
       "      <th>2012</th>\n",
       "      <th>Guinea</th>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>40</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">AG1000G-GN-B</th>\n",
       "      <th rowspan=\"2\" valign=\"top\">2012</th>\n",
       "      <th>Guinea</th>\n",
       "      <td>0</td>\n",
       "      <td>7</td>\n",
       "      <td>83</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Mali</th>\n",
       "      <td>0</td>\n",
       "      <td>28</td>\n",
       "      <td>65</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-GQ</th>\n",
       "      <th>2002</th>\n",
       "      <th>Equatorial Guinea</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-GW</th>\n",
       "      <th>2010</th>\n",
       "      <th>Guinea-Bissau</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>8</td>\n",
       "      <td>93</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">AG1000G-KE</th>\n",
       "      <th>2000</th>\n",
       "      <th>Kenya</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2007</th>\n",
       "      <th>Kenya</th>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2012</th>\n",
       "      <th>Kenya</th>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-ML-A</th>\n",
       "      <th>2014</th>\n",
       "      <th>Mali</th>\n",
       "      <td>0</td>\n",
       "      <td>27</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-ML-B</th>\n",
       "      <th>2004</th>\n",
       "      <th>Mali</th>\n",
       "      <td>2</td>\n",
       "      <td>36</td>\n",
       "      <td>33</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-MW</th>\n",
       "      <th>2015</th>\n",
       "      <th>Malawi</th>\n",
       "      <td>41</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"2\" valign=\"top\">AG1000G-MZ</th>\n",
       "      <th>2003</th>\n",
       "      <th>Mozambique</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2004</th>\n",
       "      <th>Mozambique</th>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>71</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th rowspan=\"3\" valign=\"top\">AG1000G-TZ</th>\n",
       "      <th>2012</th>\n",
       "      <th>Tanzania</th>\n",
       "      <td>87</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2013</th>\n",
       "      <th>Tanzania</th>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>10</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2015</th>\n",
       "      <th>Tanzania</th>\n",
       "      <td>137</td>\n",
       "      <td>0</td>\n",
       "      <td>32</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-UG</th>\n",
       "      <th>2012</th>\n",
       "      <th>Uganda</th>\n",
       "      <td>82</td>\n",
       "      <td>0</td>\n",
       "      <td>207</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AG1000G-X</th>\n",
       "      <th>-1</th>\n",
       "      <th>Lab Cross</th>\n",
       "      <td>0</td>\n",
       "      <td>74</td>\n",
       "      <td>49</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>174</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-ea66cd18-8f81-4b50-ac93-c2f7e71b4dc9')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-ea66cd18-8f81-4b50-ac93-c2f7e71b4dc9 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-ea66cd18-8f81-4b50-ac93-c2f7e71b4dc9');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "taxon                                                arabiensis  coluzzii  \\\n",
       "sample_set   year  country                                                  \n",
       "AG1000G-AO    2009 Angola                                     0        81   \n",
       "AG1000G-BF-A  2012 Burkina Faso                               0        82   \n",
       "AG1000G-BF-B  2014 Burkina Faso                               3        53   \n",
       "AG1000G-BF-C  2004 Burkina Faso                               0         0   \n",
       "AG1000G-CD    2015 Democratic Republic of the Congo           0         0   \n",
       "AG1000G-CF    1993 Central African Republic                   0         5   \n",
       "              1994 Central African Republic                   0        13   \n",
       "AG1000G-CI    2012 Cote d'Ivoire                              0        80   \n",
       "AG1000G-CM-A  2009 Cameroon                                   0         0   \n",
       "AG1000G-CM-B  2005 Cameroon                                   0         7   \n",
       "AG1000G-CM-C  2013 Cameroon                                   2        19   \n",
       "AG1000G-FR    2011 Mayotte                                    0         0   \n",
       "AG1000G-GA-A  2000 Gabon                                      0         0   \n",
       "AG1000G-GH    2012 Ghana                                      0        64   \n",
       "AG1000G-GM-A  2011 Gambia, The                                0         0   \n",
       "AG1000G-GM-B  2012 Gambia, The                                0         2   \n",
       "AG1000G-GM-C  2012 Gambia, The                                0        12   \n",
       "AG1000G-GN-A  2012 Guinea                                     0         4   \n",
       "AG1000G-GN-B  2012 Guinea                                     0         7   \n",
       "                   Mali                                       0        28   \n",
       "AG1000G-GQ    2002 Equatorial Guinea                          0         0   \n",
       "AG1000G-GW    2010 Guinea-Bissau                              0         0   \n",
       "AG1000G-KE    2000 Kenya                                      0         0   \n",
       "              2007 Kenya                                      3         0   \n",
       "              2012 Kenya                                     10         0   \n",
       "AG1000G-ML-A  2014 Mali                                       0        27   \n",
       "AG1000G-ML-B  2004 Mali                                       2        36   \n",
       "AG1000G-MW    2015 Malawi                                    41         0   \n",
       "AG1000G-MZ    2003 Mozambique                                 0         0   \n",
       "              2004 Mozambique                                 0         0   \n",
       "AG1000G-TZ    2012 Tanzania                                  87         0   \n",
       "              2013 Tanzania                                   1         0   \n",
       "              2015 Tanzania                                 137         0   \n",
       "AG1000G-UG    2012 Uganda                                    82         0   \n",
       "AG1000G-X    -1    Lab Cross                                  0        74   \n",
       "\n",
       "taxon                                                gambiae  gcx1  gcx2  \\\n",
       "sample_set   year  country                                                 \n",
       "AG1000G-AO    2009 Angola                                  0     0     0   \n",
       "AG1000G-BF-A  2012 Burkina Faso                           98     0     0   \n",
       "AG1000G-BF-B  2014 Burkina Faso                           46     0     0   \n",
       "AG1000G-BF-C  2004 Burkina Faso                           13     0     0   \n",
       "AG1000G-CD    2015 Democratic Republic of the Congo       76     0     0   \n",
       "AG1000G-CF    1993 Central African Republic                2     0     0   \n",
       "              1994 Central African Republic               53     0     0   \n",
       "AG1000G-CI    2012 Cote d'Ivoire                           0     0     0   \n",
       "AG1000G-CM-A  2009 Cameroon                              303     0     0   \n",
       "AG1000G-CM-B  2005 Cameroon                               90     0     0   \n",
       "AG1000G-CM-C  2013 Cameroon                               23     0     0   \n",
       "AG1000G-FR    2011 Mayotte                                23     0     0   \n",
       "AG1000G-GA-A  2000 Gabon                                  69     0     0   \n",
       "AG1000G-GH    2012 Ghana                                  36     0     0   \n",
       "AG1000G-GM-A  2011 Gambia, The                             0    68     6   \n",
       "AG1000G-GM-B  2012 Gambia, The                             0     9    19   \n",
       "AG1000G-GM-C  2012 Gambia, The                             2     0   160   \n",
       "AG1000G-GN-A  2012 Guinea                                 40     0     0   \n",
       "AG1000G-GN-B  2012 Guinea                                 83     0     0   \n",
       "                   Mali                                   65     0     0   \n",
       "AG1000G-GQ    2002 Equatorial Guinea                      10     0     0   \n",
       "AG1000G-GW    2010 Guinea-Bissau                           8    93     0   \n",
       "AG1000G-KE    2000 Kenya                                  19     0     0   \n",
       "              2007 Kenya                                   0     0     0   \n",
       "              2012 Kenya                                   0     0     0   \n",
       "AG1000G-ML-A  2014 Mali                                   33     0     0   \n",
       "AG1000G-ML-B  2004 Mali                                   33     0     0   \n",
       "AG1000G-MW    2015 Malawi                                  0     0     0   \n",
       "AG1000G-MZ    2003 Mozambique                              3     0     0   \n",
       "              2004 Mozambique                             71     0     0   \n",
       "AG1000G-TZ    2012 Tanzania                                0     0     0   \n",
       "              2013 Tanzania                               32     0     0   \n",
       "              2015 Tanzania                               32     0     0   \n",
       "AG1000G-UG    2012 Uganda                                207     0     0   \n",
       "AG1000G-X    -1    Lab Cross                              49     0     0   \n",
       "\n",
       "taxon                                                gcx3  \\\n",
       "sample_set   year  country                                  \n",
       "AG1000G-AO    2009 Angola                               0   \n",
       "AG1000G-BF-A  2012 Burkina Faso                         0   \n",
       "AG1000G-BF-B  2014 Burkina Faso                         0   \n",
       "AG1000G-BF-C  2004 Burkina Faso                         0   \n",
       "AG1000G-CD    2015 Democratic Republic of the Congo     0   \n",
       "AG1000G-CF    1993 Central African Republic             0   \n",
       "              1994 Central African Republic             0   \n",
       "AG1000G-CI    2012 Cote d'Ivoire                        0   \n",
       "AG1000G-CM-A  2009 Cameroon                             0   \n",
       "AG1000G-CM-B  2005 Cameroon                             0   \n",
       "AG1000G-CM-C  2013 Cameroon                             0   \n",
       "AG1000G-FR    2011 Mayotte                              0   \n",
       "AG1000G-GA-A  2000 Gabon                                0   \n",
       "AG1000G-GH    2012 Ghana                                0   \n",
       "AG1000G-GM-A  2011 Gambia, The                          0   \n",
       "AG1000G-GM-B  2012 Gambia, The                          0   \n",
       "AG1000G-GM-C  2012 Gambia, The                          0   \n",
       "AG1000G-GN-A  2012 Guinea                               0   \n",
       "AG1000G-GN-B  2012 Guinea                               0   \n",
       "                   Mali                                 0   \n",
       "AG1000G-GQ    2002 Equatorial Guinea                    0   \n",
       "AG1000G-GW    2010 Guinea-Bissau                        0   \n",
       "AG1000G-KE    2000 Kenya                                0   \n",
       "              2007 Kenya                                0   \n",
       "              2012 Kenya                               54   \n",
       "AG1000G-ML-A  2014 Mali                                 0   \n",
       "AG1000G-ML-B  2004 Mali                                 0   \n",
       "AG1000G-MW    2015 Malawi                               0   \n",
       "AG1000G-MZ    2003 Mozambique                           0   \n",
       "              2004 Mozambique                           0   \n",
       "AG1000G-TZ    2012 Tanzania                             0   \n",
       "              2013 Tanzania                            10   \n",
       "              2015 Tanzania                             1   \n",
       "AG1000G-UG    2012 Uganda                               0   \n",
       "AG1000G-X    -1    Lab Cross                            0   \n",
       "\n",
       "taxon                                                intermediate_gambcolu_arabiensis  \\\n",
       "sample_set   year  country                                                              \n",
       "AG1000G-AO    2009 Angola                                                           0   \n",
       "AG1000G-BF-A  2012 Burkina Faso                                                     0   \n",
       "AG1000G-BF-B  2014 Burkina Faso                                                     0   \n",
       "AG1000G-BF-C  2004 Burkina Faso                                                     0   \n",
       "AG1000G-CD    2015 Democratic Republic of the Congo                                 0   \n",
       "AG1000G-CF    1993 Central African Republic                                         0   \n",
       "              1994 Central African Republic                                         0   \n",
       "AG1000G-CI    2012 Cote d'Ivoire                                                    0   \n",
       "AG1000G-CM-A  2009 Cameroon                                                         0   \n",
       "AG1000G-CM-B  2005 Cameroon                                                         0   \n",
       "AG1000G-CM-C  2013 Cameroon                                                         0   \n",
       "AG1000G-FR    2011 Mayotte                                                          0   \n",
       "AG1000G-GA-A  2000 Gabon                                                            0   \n",
       "AG1000G-GH    2012 Ghana                                                            0   \n",
       "AG1000G-GM-A  2011 Gambia, The                                                      0   \n",
       "AG1000G-GM-B  2012 Gambia, The                                                      0   \n",
       "AG1000G-GM-C  2012 Gambia, The                                                      0   \n",
       "AG1000G-GN-A  2012 Guinea                                                           0   \n",
       "AG1000G-GN-B  2012 Guinea                                                           0   \n",
       "                   Mali                                                             0   \n",
       "AG1000G-GQ    2002 Equatorial Guinea                                                0   \n",
       "AG1000G-GW    2010 Guinea-Bissau                                                    0   \n",
       "AG1000G-KE    2000 Kenya                                                            0   \n",
       "              2007 Kenya                                                            0   \n",
       "              2012 Kenya                                                            0   \n",
       "AG1000G-ML-A  2014 Mali                                                             0   \n",
       "AG1000G-ML-B  2004 Mali                                                             0   \n",
       "AG1000G-MW    2015 Malawi                                                           0   \n",
       "AG1000G-MZ    2003 Mozambique                                                       0   \n",
       "              2004 Mozambique                                                       0   \n",
       "AG1000G-TZ    2012 Tanzania                                                         0   \n",
       "              2013 Tanzania                                                         0   \n",
       "              2015 Tanzania                                                         0   \n",
       "AG1000G-UG    2012 Uganda                                                           1   \n",
       "AG1000G-X    -1    Lab Cross                                                        0   \n",
       "\n",
       "taxon                                                intermediate_gambiae_coluzzii  \n",
       "sample_set   year  country                                                          \n",
       "AG1000G-AO    2009 Angola                                                        0  \n",
       "AG1000G-BF-A  2012 Burkina Faso                                                  1  \n",
       "AG1000G-BF-B  2014 Burkina Faso                                                  0  \n",
       "AG1000G-BF-C  2004 Burkina Faso                                                  0  \n",
       "AG1000G-CD    2015 Democratic Republic of the Congo                              0  \n",
       "AG1000G-CF    1993 Central African Republic                                      0  \n",
       "              1994 Central African Republic                                      0  \n",
       "AG1000G-CI    2012 Cote d'Ivoire                                                 0  \n",
       "AG1000G-CM-A  2009 Cameroon                                                      0  \n",
       "AG1000G-CM-B  2005 Cameroon                                                      0  \n",
       "AG1000G-CM-C  2013 Cameroon                                                      0  \n",
       "AG1000G-FR    2011 Mayotte                                                       0  \n",
       "AG1000G-GA-A  2000 Gabon                                                         0  \n",
       "AG1000G-GH    2012 Ghana                                                         0  \n",
       "AG1000G-GM-A  2011 Gambia, The                                                   0  \n",
       "AG1000G-GM-B  2012 Gambia, The                                                   1  \n",
       "AG1000G-GM-C  2012 Gambia, The                                                   0  \n",
       "AG1000G-GN-A  2012 Guinea                                                        1  \n",
       "AG1000G-GN-B  2012 Guinea                                                        1  \n",
       "                   Mali                                                          1  \n",
       "AG1000G-GQ    2002 Equatorial Guinea                                             0  \n",
       "AG1000G-GW    2010 Guinea-Bissau                                                 0  \n",
       "AG1000G-KE    2000 Kenya                                                         0  \n",
       "              2007 Kenya                                                         0  \n",
       "              2012 Kenya                                                         0  \n",
       "AG1000G-ML-A  2014 Mali                                                          0  \n",
       "AG1000G-ML-B  2004 Mali                                                          0  \n",
       "AG1000G-MW    2015 Malawi                                                        0  \n",
       "AG1000G-MZ    2003 Mozambique                                                    0  \n",
       "              2004 Mozambique                                                    0  \n",
       "AG1000G-TZ    2012 Tanzania                                                      0  \n",
       "              2013 Tanzania                                                      0  \n",
       "              2015 Tanzania                                                      0  \n",
       "AG1000G-UG    2012 Uganda                                                        0  \n",
       "AG1000G-X    -1    Lab Cross                                                   174  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pivot_country_year_taxon = (\n",
    "    metadata\n",
    "    .pivot_table(\n",
    "        index=[\"sample_set\", \"year\", \"country\"], \n",
    "        columns=[\"taxon\"], \n",
    "        values=\"sample_id\",\n",
    "        aggfunc=\"count\",\n",
    "        fill_value=0\n",
    "    )\n",
    ")\n",
    "\n",
    "pivot_country_year_taxon"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "BLwvXBT28NYE",
   "metadata": {
    "id": "BLwvXBT28NYE"
   },
   "source": [
    "Here, we can see the breakdown by sample set for country, species and year. For the purposes of this notebook, let's use the Ghana sample set. If we wanted to use all sample sets, we could supply '3.0' instead of a sample set, which will load all samples from the ag3.0 release."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "49ca7f2b",
   "metadata": {
    "id": "49ca7f2b"
   },
   "outputs": [],
   "source": [
    "sample_set = 'AG1000G-UG'          # sample_set = '3.0' .you can also supply lists with multiple sample sets e.g ['AG1000G-GH', 'AG1000G-CI', 'AG1000G-BF-A]\n",
    "sample_query = \"aim_species == 'coluzzii'\"                # here we can subset to specific values in the metadata e.g :     \"aim_species == 'gambiae'\" , or \"aim_species == 'arabiensis'\"          "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67coxQKoo82x",
   "metadata": {
    "id": "67coxQKoo82x"
   },
   "source": [
    "### **Plot allele frequencies in primers locations**\n",
    "\n",
    "Now we can plot the primers pairs, and the frequency of any alternate alleles in the Ag1000g sample set of choice. When calculating allele frequencies, we will take the sum of all alternate alleles, as we are interested here in any mutations which are different from the reference genome. We can see the frequencies of specific alleles by hovering over the points of the plot - in some cases it may be preferable to design degenerate primers rather than avoid a primer pair completely.\n",
    "\n",
    "We will also plot the primer Tm, GC and genomic spans of each primer binding site. We can use this plot to identify primers pairs and probes which may be suitable, particularly trying to avoid SNPs in the 3' end. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "shXKBpG49LSU",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 1000
    },
    "id": "shXKBpG49LSU",
    "outputId": "3d6287e4-2ce5-4b18-b213-5fa956f8f189"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue Aug 30 21:27:59 2022 DEBUG snp_calls: normalise parameters\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Subsetting allele frequencies to aim_species == 'coluzzii'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Tue Aug 30 21:27:59 2022 DEBUG snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
      "Tue Aug 30 21:27:59 2022 DEBUG _snp_calls_dataset: call arrays\n",
      "Tue Aug 30 21:28:00 2022 DEBUG _snp_calls_dataset: sample arrays\n",
      "Tue Aug 30 21:28:00 2022 DEBUG _snp_calls_dataset: create a dataset\n",
      "Tue Aug 30 21:28:00 2022 DEBUG snp_calls: concatenate data from multiple sample sets\n",
      "Tue Aug 30 21:28:00 2022 DEBUG snp_calls: add variants variables\n",
      "Tue Aug 30 21:28:00 2022 DEBUG _snp_variants_dataset: variant arrays\n",
      "Tue Aug 30 21:28:01 2022 DEBUG _snp_variants_dataset: variant_position\n",
      "Tue Aug 30 21:28:01 2022 DEBUG _snp_variants_dataset: variant_allele\n",
      "Tue Aug 30 21:28:01 2022 DEBUG _snp_variants_dataset: variant_contig\n",
      "Tue Aug 30 21:28:01 2022 DEBUG _snp_variants_dataset: site filters arrays\n",
      "Tue Aug 30 21:28:02 2022 DEBUG _snp_variants_dataset: set up attributes\n",
      "Tue Aug 30 21:28:02 2022 DEBUG _snp_variants_dataset: create a dataset\n",
      "Tue Aug 30 21:28:02 2022 DEBUG snp_calls: handle site class\n",
      "Tue Aug 30 21:28:02 2022 DEBUG snp_calls: handle region, do this only once - optimisation\n",
      "Tue Aug 30 21:28:05 2022 DEBUG snp_calls: concatenate data from multiple regions\n",
      "Tue Aug 30 21:28:05 2022 DEBUG snp_calls: apply site filters\n",
      "Tue Aug 30 21:28:05 2022 DEBUG snp_calls: add call_genotype_mask\n",
      "Tue Aug 30 21:28:05 2022 DEBUG snp_calls: handle sample query\n",
      "Tue Aug 30 21:28:05 2022 DEBUG snp_calls: handle cohort size\n",
      "Tue Aug 30 21:28:08 2022 DEBUG snp_calls: normalise parameters\n",
      "Tue Aug 30 21:28:08 2022 DEBUG snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_calls_dataset: call arrays\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_calls_dataset: sample arrays\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_calls_dataset: create a dataset\n",
      "Tue Aug 30 21:28:08 2022 DEBUG snp_calls: concatenate data from multiple sample sets\n",
      "Tue Aug 30 21:28:08 2022 DEBUG snp_calls: add variants variables\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_variants_dataset: variant arrays\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_variants_dataset: variant_position\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_variants_dataset: variant_allele\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_variants_dataset: variant_contig\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_variants_dataset: site filters arrays\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_variants_dataset: set up attributes\n",
      "Tue Aug 30 21:28:08 2022 DEBUG _snp_variants_dataset: create a dataset\n",
      "Tue Aug 30 21:28:08 2022 DEBUG snp_calls: handle site class\n",
      "Tue Aug 30 21:28:08 2022 DEBUG snp_calls: handle region, do this only once - optimisation\n",
      "Tue Aug 30 21:28:11 2022 DEBUG snp_calls: concatenate data from multiple regions\n",
      "Tue Aug 30 21:28:11 2022 DEBUG snp_calls: apply site filters\n",
      "Tue Aug 30 21:28:11 2022 DEBUG snp_calls: add call_genotype_mask\n",
      "Tue Aug 30 21:28:11 2022 DEBUG snp_calls: handle sample query\n",
      "Tue Aug 30 21:28:11 2022 DEBUG snp_calls: handle cohort size\n",
      "Tue Aug 30 21:28:14 2022 DEBUG snp_calls: normalise parameters\n",
      "Tue Aug 30 21:28:14 2022 DEBUG snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_calls_dataset: call arrays\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_calls_dataset: sample arrays\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_calls_dataset: create a dataset\n",
      "Tue Aug 30 21:28:14 2022 DEBUG snp_calls: concatenate data from multiple sample sets\n",
      "Tue Aug 30 21:28:14 2022 DEBUG snp_calls: add variants variables\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_variants_dataset: variant arrays\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_variants_dataset: variant_position\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_variants_dataset: variant_allele\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_variants_dataset: variant_contig\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_variants_dataset: site filters arrays\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_variants_dataset: set up attributes\n",
      "Tue Aug 30 21:28:14 2022 DEBUG _snp_variants_dataset: create a dataset\n",
      "Tue Aug 30 21:28:14 2022 DEBUG snp_calls: handle site class\n",
      "Tue Aug 30 21:28:14 2022 DEBUG snp_calls: handle region, do this only once - optimisation\n",
      "Tue Aug 30 21:28:17 2022 DEBUG snp_calls: concatenate data from multiple regions\n",
      "Tue Aug 30 21:28:17 2022 DEBUG snp_calls: apply site filters\n",
      "Tue Aug 30 21:28:17 2022 DEBUG snp_calls: add call_genotype_mask\n",
      "Tue Aug 30 21:28:17 2022 DEBUG snp_calls: handle sample query\n",
      "Tue Aug 30 21:28:17 2022 DEBUG snp_calls: handle cohort size\n",
      "Tue Aug 30 21:28:20 2022 DEBUG snp_calls: normalise parameters\n",
      "Tue Aug 30 21:28:20 2022 DEBUG snp_calls: access SNP calls and concatenate multiple sample sets and/or regions\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_calls_dataset: call arrays\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_calls_dataset: sample arrays\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_calls_dataset: create a dataset\n",
      "Tue Aug 30 21:28:20 2022 DEBUG snp_calls: concatenate data from multiple sample sets\n",
      "Tue Aug 30 21:28:20 2022 DEBUG snp_calls: add variants variables\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_variants_dataset: variant arrays\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_variants_dataset: variant_position\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_variants_dataset: variant_allele\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_variants_dataset: variant_contig\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_variants_dataset: site filters arrays\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_variants_dataset: set up attributes\n",
      "Tue Aug 30 21:28:20 2022 DEBUG _snp_variants_dataset: create a dataset\n",
      "Tue Aug 30 21:28:20 2022 DEBUG snp_calls: handle site class\n",
      "Tue Aug 30 21:28:20 2022 DEBUG snp_calls: handle region, do this only once - optimisation\n",
      "Tue Aug 30 21:28:22 2022 DEBUG snp_calls: concatenate data from multiple regions\n",
      "Tue Aug 30 21:28:22 2022 DEBUG snp_calls: apply site filters\n",
      "Tue Aug 30 21:28:22 2022 DEBUG snp_calls: add call_genotype_mask\n",
      "Tue Aug 30 21:28:22 2022 DEBUG snp_calls: handle sample query\n",
      "Tue Aug 30 21:28:22 2022 DEBUG snp_calls: handle cohort size\n"
     ]
    },
    {
     "ename": "AttributeError",
     "evalue": "'numpy.ndarray' object has no attribute 'compute'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_7498/2408643155.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      6\u001b[0m                                                 \u001b[0massay_type\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0massay_type\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m                                                 \u001b[0mseq_parameters\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mseq_parameters\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m                                                 save=True)\n\u001b[0m",
      "\u001b[0;32m~/projects/AgamPrimer/AgamPrimer/AgamPrimer.py\u001b[0m in \u001b[0;36mplot_primer_ag3_frequencies\u001b[0;34m(primer_df, gdna_pos, contig, sample_set, assay_type, seq_parameters, save, sample_query)\u001b[0m\n\u001b[1;32m    342\u001b[0m   \u001b[0;31m# Loop through each primer pair and get the frequencies of alternate alleles, storing in dict\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    343\u001b[0m   \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mn_primer_pairs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 344\u001b[0;31m     \u001b[0mres_dict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mget_primer_alt_frequencies\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mprimer_df\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgdna_pos\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mi\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_set\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0massay_type\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcontig\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msample_query\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    345\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    346\u001b[0m   \u001b[0;31m# Plot data with plotly\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/projects/AgamPrimer/AgamPrimer/AgamPrimer.py\u001b[0m in \u001b[0;36mget_primer_alt_frequencies\u001b[0;34m(primer_df, gdna_pos, pair, sample_set, assay_type, contig, sample_query)\u001b[0m\n\u001b[1;32m    233\u001b[0m       \u001b[0mfreq\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfreq_arr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mprimer_size\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    234\u001b[0m       \u001b[0mbase_freqs_arr\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mbase_freqs\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mprimer_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 235\u001b[0;31m       \u001b[0mref\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mref_alt_arr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mprimer_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    236\u001b[0m       \u001b[0mref_alt\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mref_alt_arr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mprimer_size\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m:\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcompute\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    237\u001b[0m       \u001b[0mpos\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mpos_arr\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0mprimer_loc\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0mprimer_size\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'numpy.ndarray' object has no attribute 'compute'"
     ]
    }
   ],
   "source": [
    "results_dict = AgamPrimer.plot_primer_ag3_frequencies(primer_df=primer_df,\n",
    "                                                gdna_pos=gdna_pos,\n",
    "                                                contig=contig,\n",
    "                                                sample_set=sample_set, \n",
    "                                                sample_query=sample_query,\n",
    "                                                assay_type=assay_type,\n",
    "                                                seq_parameters=seq_parameters,\n",
    "                                                save=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aCVuE1CiAR4h",
   "metadata": {
    "id": "aCVuE1CiAR4h"
   },
   "source": [
    "Now lets plot these primer pairs across the genome, highlighting where they are in relation to any nearby exons..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "cAacbw-q8Sgm",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 302
    },
    "id": "cAacbw-q8Sgm",
    "outputId": "879684f4-91eb-4e59-e8d1-e31c8397b0ba"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "DEBUG:malariagen_data.ag3:geneset: handle region\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA5kAAAENCAYAAACFPqFaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4yLjIsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+WH4yJAAAgAElEQVR4nO3df5xWY/748dc1M22plPRDfkTyoya10kpWIWStNflV1CrkR1GyWCUsoV35sfiwVLsWyzcSFWpr1y6KLdtSIk2Sn6MSVkmaalLT9f3jvmunaaYm3dM907yej8d5dJ9znXOd9313zZl539d1nRNijEiSJEmSlAoZ6Q5AkiRJkrTrMMmUJEmSJKWMSaYkSZIkKWVMMiVJkiRJKWOSKUmSJElKmaxyqtdb1kqSJEnSriuUVmBPpiRJkiQpZUwyJUmSJEkpY5IpSZIkSUoZk0xJkiRJUsqYZEqSJEmSUsYkU5IkSZKUMiaZkiRJkqSUMcmUJEmSJKWMSaYkSZIkKWVMMiVJkiRJKWOSKUmSJElKGZNMSZIkSVLKmGRKkiRJklLGJFOSJEmSlDImmZIkSZKklDHJlCRJkiSljEmmJEmSJCllTDIlSZIkSSljkilJkiRJShmTTEmSJElSyphkSpIkSZJSxiRTkiRJkpQyJpmSJEmSpJQxyZQkSZIkpYxJpiRJkiQpZUwyJUmSJEkpY5IpSZIkSUoZk0xJFdunZ8IHbXasjg/awLrPUxOPJEmStsokU1LFtuG7dEcgSZKk7ZCV7gCktFu/zESmIiv8NjX1rFsM8ftiG7PgR01SU78kSZIAk0wJPs0BAoSQ7khUog2QWX/Hq/n8KgiZxapeC00eg91+vOP1S5IkCTDJlCAWABFiugNRqWodu4MVZCT+n4v/H2fUhg35O1i3JEmSinJOpqRKwF5mSZKkysKeTCmjFsT16Y5CpYkFkP8KcMsOVLIBqFbCcNk1kFl3B+qVJElScSaZ0oF/hw2r0x2FSpN3RmpuzNTkMchqtPm2kAVZKZjvKQmA5z66hFb1u3FovVO3KBv/4UW0btCdQ+v9PA2RSZJ2JpNMKXP3xKKKKWM3KCzY8Xqy9oRqe+14PZJKtbbwO5au+bDEJPP7DfksX/tJGqKSJO1sJpmSKrasvSGuS0FFTkGXdobIhlLLgj+HklQlmGRKqtgOGL3jdRz6zo7XIWmTKYuGsqzgoxLLqmfWLvW4j1e8wmcr/11eYZXoJ40uolndE3bqOSWpqjPJlCRJ22V5wSes37Bmi+2Z4Uc02K15icfsUb0p367NK/G48vTt2oU79XySJMePSZKkFAlkEOPmD6SNMfLlva+T4Z8cklRl2JMpSZK2SwgZZIbqhCLPsI3fF7K+WgFrC1dsvvOGCBsi36z9BAhkheo7Lc4Ncf1mMUqSdg6TTEmStF2O3+83rPx+yWbbvp20gPk/+ceWczWLdGzu/qO9aVW/206I8H/2rnn4Tj2fJMkkU5IkldGGNev4duL7xLWF1GbzRz/V+KoF83/yjy3vIFtk+OzuixtSe9rOeWRUtX12p27ng3bKuSRJmzPJ1C7lvidmcX6XljTcs2aZ9h/7jwU0abw7Rx++TzlHJkmVX+GKAtYtWUlcv5XHlITNh6cWnaK5oWAd67/KL6/wNlP47RqTTElKE5NM7XJWrvq+zEnmoi9Xsm79BpNMSSqrjFLmOGYETt33XnbbrV6JZe2m9KT62lrlG5skqUIwydQup9iNDbcpeE8ISSqzWFhKL+aGSN3q+xIyMzfbHEKirHZ+g/IPTpJUIZhkqtJZX7iBMX97n1Vr1pVYnpm5fVnjF1+v4k/PzklFaEqx+P0a2tf4nCO6dd3qfoX39yWjxw2ExgfupMikqimrQS1qtd+P+P2WiebqWZ9v9Vu7mkfuW56hbSGrkb2mkpQuJpmqdNYXbmDpt2vYsGHLLsuMEKi1W7Uy19V0nzrkLfmu1IRV6ZbFF8sKOGJbu20oJK742iRTKmchK4PdOxxQYtnqWZ+XPJQ2mXjWOcGfT0mqKnwysiqpksfEZmWFUkpKllHa3CJVOsVvNiIpDUr6MfQ6K0lVjj2ZqpQKN0QikWrF5v58v24DcTsmZf73q28gZpBJYapDVArE7XiI+oZJf4Tdn9l8Y8ggo0t/wl5NUxuYpC3s1qZxyV/2hESZJKnqMMlUpVPjR1l8t+8yCgoK+eXB7TYrm/LGQlau+p696pdtLk7+ukwIcGz+v8sjVKVA0yYl3KmyuKwfwfrvYeXyYturEb9eZJIp7QR1Tz64xO0hhFLLJEm7pnJPMmOMHHjggXz22WcAvPfee2RnZ28q//zzz/nd737Hiy++yJIlS6hduzbNmjXj7LPP5oYbbtisruOOO45p06YBMHnyZH7xi19sKrv11lu57bbbNq3XqFGDZs2acfnll3PllVdu2j58+HDuu+8+Fi9eTNOmTbnxxhu58MILN5XPmTOHq6++mv/85z/UrFmTs88+m/vuu4/dd//fw6Off/55hg0bRm5uLj/60Y9o3bo1f/3rX6lXrx6jR49m5MiRvP/++6xatYpmzZpxzTXXcMkllwDw6aefMmjQIGbNmsWXX35J/fr1Oe2007jrrruoVy/xx3ReXh7XXnst06ZN47vvvmO//fbjggsu4Oabb65yQwK/Wv0d73/71Rbbv6z+DVSHNi0abbZ9yhsLt/szarzuKw5fO2+H4lT5+WrdYSwrWEX9Glv54qBa9USSuYWq9fMiSZJUEZR7kjlt2rRNCSbAqFGjGDZsGAAffPABHTp0YOnSpTRs2JCuXbtSrVo13n33Xe67777NkszPPvuM6dOnb1p/8sknN0syN2rWrBldunRh8eLFPPfcc/zqV7+iRo0a9OnThzFjxjBgwAAaNmzIL3/5SyZOnEjv3r1p3Lgxp5xyCitXruTkk0/m66+/pmvXrnz66ac88sgj5Ofn8/TTTwPw9NNPc95551G9enXOPPNMateuzZtvvsnq1aupV68e//znP8nLy+OUU07hv//9Ly+99BKXXnopjRo1okuXLnz22WdMmjSJE044gRNPPJGxY8fy5z//mWXLljF+/HgAevfuzWuvvUaLFi3o0qULo0eP5pZbbqFJkyZcdNFF5fL/VFGN/fRt3lv+BRklJI5NapXcw7W9efiGEFhXxZL3yiIrRgqWLuLlhXO54NCjS99xzUrIyIBQbJp5LP2B8ZIkSSof5Z5kPvnkkwAcccQRvP3224wePZrbb7+dEAJXXXUVS5cupXnz5vz73/9mzz333HTc3LlzN6vnqaeeIsa4qZ4JEyaQn59P7dq1N9uvdevW3H///QB06dKFSZMm8be//Y0+ffpw5513AjBy5Ei6du3Ko48+yqWXXsodd9zBKaecwqOPPsrXX39NTk4O48aNIz8/n4YNG/Lss89y++23c+CBBzJ48GAAXnzxRTp16rTF+7366qt59NFHyUzOFezUqROvvfYaL730El26dOHQQw/l008/Ze+99wYSvbMXXXQRL7300qY6Pv74YwB+//vfk5OTw4oVKxg/fjx5eXk/9L+h0ooxUhg3UFjCNMt9atXdYtsZJx7Mvo1233LnUnQ/pj5z57/LC3UcylURnbPoQw5Y9R3TyjLN9sDDCU2aF9sYCIceWR6hSZIkqRTlmmSuXbuWcePGAXDvvffStWtXPvvsM/71r39x1FFH8fLLLwOJxKxoggmJZLGop556CoBf//rX3HnnncybN4/x48dvNtS1qM8//5wFCxYA0KBBA9avX09ubi4ARx555Gb/vvPOOwC8/fbbm22vXbs2LVq04J133uHdd99l/fr1LFq0iN122427776bnJwcGjduzDXXXMMVV1wBQJs2bTaL4/vvE0P49ttvPwD22WefrZYDXHvttVx77bUMGjSIF154gcmTJ3PAAQfQu3fvEt9rVVXSDX4OarLHdtWx7yEH8mzBJ7z/bY1UhaUU6vzlQuqtW1umfUOL9mQ0b7ftHSVJklSuyjXJnDx5MsuXL6dRo0Ycf/zx5OTkMGrUKJ588kkOPvhg1q9fD0DTpk2BRO/gqaeeuun4qVOn0qlTJ2bPns17771HVlYWOTk5fPDBB8ybN48nn3xyiyRzwoQJm83J23PPPRk4cCBLly6lsDBxB9GNvZ+1aiXmeK1YsYKCggK++uqrzcqL7vPll1/SqFFi/t+aNWv45JNPOPfcc3n66acZMGAA++67L2eeeeZmsdx3333MmDGDgw8+mMsvv3yLz2fBggX85je/ISMjg7vvvnvT9pNOOonWrVszZ84c3n//fQBOO+00GjeuenfnO6RuIxbmf7PF9u/WFfDm159xSYsOO3yOjXM+61Qz0axoNj75oFndBtve2aGxkiRJFUK5Jpkbh8p26dKFjIwMzjrrLEaNGsXYsWO55557yMrK2tQ7CIlk86qrrmLkyJGbeviK1nP88cezxx57cNZZZ/Hb3/6WKVOmsGTJks16BzfOyaxZsybNmjXjnHPOoW7duqxfv57MzEwKCwvJz8+nfv365OfnA1C3bl1q1KjBXnvtBbBpe9HXjRs3pmHDhpu2jxo1inbt2rHbbrsxYsQIJk6cuFmSufFGRM2aNeOVV16hTp06m302M2fO5Be/+AXLly/nscceIycnB4DCwkJ+/vOfs2TJEp566ilOP/10evbsyYgRI6hVq9ZmyWhV8PMmLfl5k5ZbbL/+zRdYvnZ1ys5zSJ1GDDy8c8rqU2rEg46CdWs5tvFBW90v5PQjNGmxk6KSJEnS1mRse5cf5ttvv+Vvf/sbAI8++ighBM4++2wg0XP40ksvcdJJJwHw4IMPsnLlSlq0aMH999/PbrvttqmewsJCxowZA8Arr7xCCIG2bdsCsGHDBkaPHr3ZeTfOyRw2bBiXXnopdesm5u1lZWVx2GGHAfDmm28CiUQP4PDDDwf+N9R1Y/nKlSt5//33CSHQunVrDjjggC2SxY029n5u2LCB/v37c9ttt3HEEUfw+uuvs//++2+270svvcSJJ55Ifn7+FkN+ly9fzpIlSwD46U9/Su3atTfFNX/+/K1/6FXI9jwLs0z1kdr6lBqh4X6EfbaeYAJkHHokYbfa29xPkiRJ5a/cejKfffZZ1q5dS506dTjhhBM2bX/vvff48MMPGTVqFPfffz8dOnRg7ty5ZGdnc9JJJxFCYNWqVZv2f+WVV/jiiy/IysritNNO27Q9Ly+POXPmMGrUKAYOHFimmAYPHkzPnj254oormDx5MhMmTADg+uuvB+DSSy/l9ttv529/+xvdunXjk08+Ye3atZx77rkcdFDiD92rr76aoUOHcsEFF/DTn/6Up59+mszMTHr27AnAzTffzMiRI8nIyOCII47YdLOhgw8+mAEDBjBv3jxycnL4/vvvOeaYY5g6dSpTp04FYMiQITRo0IBDDz2UDz74gDPPPJOjjjqKZ55JPGC+Y8eOP+j/Yld0ZMMDWFqQv+0dy+DgOg1ptec+295RkiRJ0rbFGMtjiccee2wE4qBBg2JRr776agRitWrV4tKlS2NeXl685JJL4r777hurVasWGzZsGNu3bx+HDRsWly9fHs8///wIxHPOOWezej755JOYkZERgfjuu+/GW265JQLxjDPOiFvzwAMPxAMPPDBWq1YtHnLIIfHRRx/drHz27Nnx+OOPjzVq1Ij16tWLF198cVyxYsWm8nXr1sXrr78+Nm7cONasWTMeeeSRcfLkyZvKL7zwwghssRx//PExxhinTp1aYjkQP/300xhjjO+//34888wzY6NGjWL16tVjs2bN4k033RTXr1+/1fcmSZIkSTtJqflgiCkedrgxdy2PSiVJkiRJFUKpD5ovtzmZkiRJkqSqxyRTkiRJkpQyJpmSJEmSpJQxyZQkSZIkpYxJpiRJkiQpZUwyJUmSJEkpY5IpSZIkSUoZk0xJkiRJUspkpTuAXVUIpT6bVJIkSVIlF2NMdwgVlj2ZkiRJkqSUMcmUJEmSJKWMw2V3gnvuuSfdIagMrr322nSHoKR777033SFst53Rfirj5yIVV54/K/6MSCpvAwcOTHcIlYI9mZIkSZKklDHJlCRJkiSljEmmJEmSJCllTDIlSZIkSSljkilJkiRJShmTTEmSJElSyphkSpIkSZJSxiRTkiRJkpQyJpmSJEmSpJQxyZQkSZIkpYxJpiRJkiQpZUwyJUmSJEkpk5XuAFQ5DRo0iL333nvTeps2bTjxxBOZPn0606ZNY9myZdx2223UqlUrjVFKWyqt7T711FMsXryYjIwM9t9/f7p160ZmZmYaI1W6lNZGnn32WRYtWgRAgwYN6NGjB9WrV09XmJJSqLSf+41eeOEF3nzzTYYNG5aO8KRKxyRTP0i1atX49a9/vcX2Aw88kJYtWzJy5Mg0RCVtW2ltt23btpx33nkAPPXUU7zxxhscc8wxOzs8VQCltZHTTz+dGjVqADBx4kRef/31zf4IlVR5lfZzD7Bo0SJWr169kyOSKjeHyyql9t13X/bcc890hyFtt+zsbEIIhBDYf//9WbFiRbpDUgWzMcGMMbJu3bo0RyNpZ9iwYQOTJk0iJycn3aFIlYo9mfpB1q1bx3333bdp/cQTT6RNmzZpjEgqm2213cLCQt566y3OOOOMdISnCmBrbWTMmDG8//777LXXXnTp0iVdIUpKsdJ+7l9//XUOO+ww6tSpk8bopMrHJFM/yNaGlUgV2bba7nPPPUezZs1o1qzZToxKFcnW2kiPHj3YsGEDzz//PO+88w5HHXXUTo5OUnko6ed+xYoVzJkzh379+qUpKqnycrisJCX985//JD8/3x4qbVVGRgZt2rRh7ty56Q5FUjn6/PPPWbZsGXfeeSe3334769at44477kh3WFKlYE+mJAFvvPEGCxYs4PLLLycjw+/ftLkYI8uWLaNBgwbEGHnvvfdo1KhRusOSVI5atmzJLbfcsmn9xhtv5IYbbkhjRFLlYZKpH6T43IXmzZtz2mmnMW3aNF599VVWrlzJvffeS4sWLTj33HPTGKm0udLa7vjx46lXrx4PPvggAK1ateJnP/tZusJUGpXURk499VTGjBlDQUEBMUb22WcfunbtmsYoJaVSab8bJP0wIcZYHvWWS6WVSQhh0+t77rknjZGorK699tp0h6Cke++9N90hbLed0X4q4+ciFVeePyv+jEgqbwMHDtz0upzyqMoklFbgmDBJkiRJUsqYZEqSJEmSUsYkU5IkSZKUMmVKMkMId4cQ6oQQqoUQXgkhfB1C6FXewUmSJEmSKpey9mT+LMb4HZAD5AEHA4PKKyhJkiRJUuVU1iRz46NOTgPGxhhXlFM8kiRJkqRKrKzPyZwUQngfWAP0CyE0BArKLyxJkiRJUmVUpp7MGOP1wDHAkTHGdcAq4IzyDEySJEmSVPmEsj5ENIRwDNCUIr2fMcb/V8ruPpk0lPpsUkmSJEmVXFnzqF1YqQlPmYbLhhBGAQcB7wCFyc0RKC3JlCRJkiRVQWWdk3kk0DKarkuSJEmStqKsSWYu0Bj4ohxj2aWYj0uSJEmqisqaZDYA3gshvAms3bgxxnh6uUQlSZIkSaqUyppk3lqeQUiSJEmSdg3bc3fZvYB2ydU3Y4z/LW3f0aNHx9GjRwMwatQohg4dyocffshBBx3E0KFD6dmzJwDdu3dnn3324f/+7/8AuOeee5gwYQLTpk2jdu3ajBkzhnPOOYc1a9bQqVMnTj31VAYPHgzAwIEDycvLY9y4cQA888wzDB48mLy8PLKzsxk8eDC9e/cG4Pzzz6du3bo89NBDADzwwAOMHj2aN954gz333JP/9//+H2eccQaFhYX87Gc/49hjj+Xmm28G4MYbb2TevHlMmDABgOeee44BAwawZMkSDj/8cK644gr69u0LwCWXXEJmZiYPP/wwACNHjuTPf/4zs2fPZq+99uLhhx/mjDMST3457bTTaNu2Lb/97W8BuOWWW3jzzTf5+9//DsCkSZO46KKL+PrrrznyyCO56KKLuOKKKwC4/PLLWbt2LX/5y18AeOSRR3jggQeYO3cu++23H/fffz/dunUD4Mwzz6RFixbceeedAPzud7/j1Vdf5eWXX6ZatWo8//zz9OrVi2+//Zaf/vSndO/enauvvhqAK6+8kuXLl/Pkk08C8MQTTzBs2DAWLFhAs2bNuP322/nlL38JwLnnnkuTJk249957Abj77ruZPHkyr732GjVr1uTZZ5/l3HPPZfXq1Rx//PGcdtppXHfddQBce+21LFq0iGeffRaAp59+mt/85jd88sknNG/enBtvvJELL7wQgF69elGvXj0efPBBAO6//36eeeYZZsyYwR577MGTTz7JWWedxbp16+jcuTOdOnXipptuAuD666/n/fff54UXXgBg3LhxXH311SxevJjWrVtz1VVXcemllwJw0UUXUb16df74xz8CMHz4cP7yl78wa9YsGjZsyF/+8hdycnIAOPXUUznqqKO47bbbALj55puZPXs2kydPBmDChAn07duXr776irZt29KnTx/69esHQN++fSksLOTRRx8F4OGHH2b48OHMmTOHffbZh4ceeoizzz4bgDPOOIPDDjuMYcOGAfDb3/6WadOm8c9//pPMzEwmTJjABRdcwDfffEP79u0577zzuOqqqwAYMGAAK1asYNSoUQA8/vjj3HXXXcyfP5+mTZty11130b17dwC6detG06ZNueeeewC46667+Pvf/86rr77KbrvtxtixY+nRowf5+fkce+yxnHHGGQwcOBCAa665hiVLlvDMM88A8NRTTzFkyBA+/vhjDjnkEIYMGcL5558PwHnnnUfDhg154IEHALjvvvsYP348r7/+OnXq1GH06NF07dqVtWvXcuKJJ3LyySdzww03AHDdddfx0Ucf8dxzzwHw7LPPMnDgQBYuXEjLli0ZOHAgF198MQAXXnghtWrVYsSIEQA8+OCDjBo1ijfffJP69evzxBNPcPrpp7NhwwZOOeUUjjnmGG655RYAbrrpJubMmcNf//pXAF544QX69evHF198QZs2bejXrx+XXXYZwKa288gjjwDwpz/9iZEjR/LOO++w9957M3LkSM4880wAunTpwuGHH87vfvc7AG677Tb+/e9/849//IOMjAwmTpzIhRdeyLJlyzjqqKM4//zzufLKKwHo378/q1at4oknngDgscce45577uG9995j//3355577uHcc88F4Oyzz+bggw/m7rvvBuCOO+7gpZdeYsqUKVSvXp3x48dz3nnn8d1339GhQwe6du3Kr3/9awCuuuoqvv76a7yWey33Wu613Gu513Kv5V7LS7qWT5o0qdS7y5YpyQwhnAv8HniVxK1qjwUGxRjHlXKIExIlSZIkade1w0nmHODkjb2XIYSGwMsxxsNLOcQkU5IkSZJ2XaUmmRllrCCj2PDYZdtxrCRJkiSpiijrjX9eDCH8A3g6ud4d+Hv5hCRJkiRJqqy258Y/XYEOydVpMcbnt7K7w2UlSZIkade1w3MyL4kxPlpkPRO4KcZ4WymHmGRKkiRJ0q5rh+dknhRC+FsIYe8QwmHAf4DdUxKaJEmSJGmXsT3DZbsDw4FVwHkxxte3srs9mZIkSZK069qxnswQwiHAVcB44DPg/BBCzdTEJkmSJEnaVZR1uOxfgZtjjJcBxwMfAjPLLSpJkiRJUqVU1hv/1Ikxflds26Exxg9KOWSLStetW8fixYspKCj4YZFWcjVq1GC//fajWrVq6Q5FkiRJknbUD7u7bAjhuhjj3cnX58QYxxYpGxZjvLGUQ7eo9NNPP2X33Xenfv36hFBqPLukGCPLli1j5cqVHHjggekOR5IkSZJ21A+ek9mjyOsbipX9fHsiKCgoqJIJJkAIgfr161fZXlxJkiRJVce2ksxQyuuS1repKiaYG1Xl9y5JkiSp6thWkhlLeV3SuiRJkiSpisvaRvnhIYTvSPRa7pZ8TXK9xo6ceNB/nuO7dakbPlqnWg1+f/TZKasPYMiQIRx33HF07tx5q/utXbuWCy64gLfeeov69evzzDPP0LRp05TGIkmSJEmVwVaTzBhjZnmdOJUJZnnUBzB06NAStxcWFpKZ+b+P5tFHH6VevXp89NFHjBkzhsGDB/PMM8+kPB5JkiRJqujK+pzMSi8vL48WLVrQs2dPsrOz6datG6tXrwYSyWS7du1o1aoVffv2ZeMdd3v37s24ceMAaNq0KYMHD6Zt27aMHTt2s7onTJjAhRdeCEC3bt145ZVXKMujYSRJkiRpV1NlkkyABQsW0L9/f+bPn0+dOnUYMWIEAAMGDGDmzJnk5uayZs0aJk2aVOLx9evXZ/bs2fTo0WOz7Z9//jlNmjQBICsri7p167Js2bLyfTOSJEmSVAFVqSSzSZMmdOjQAYBevXoxffp0AKZOnUr79u1p3bo1U6ZMYd68eSUe3717950WqyRJkiRVRlUqySz+GJEQAgUFBfTv359x48Yxd+5c+vTpU+rzLGvVqlXi9n333ZdFixYBsH79elasWEH9+vVTG7wkSZIkVQJVKslcuHAhM2bMAGD06NF07NhxU0LZoEED8vPzN83B3B6nn346TzzxBADjxo3jxBNP9LmYkiRJkqqkbT3CpNzUqVYj5Y8w2ZbmzZszfPhwLr74Ylq2bEm/fv2oWbMmffr0oVWrVjRu3Jh27dpt97kvueQSzj//fA4++GD23HNPxowZ80PegiRJkiRVeqGc7oK6RaXz588nOzu7PM5VJnl5eeTk5JCbm5u2GNL9GUiSJElSipQ6dLNKDZeVJEmSJJWvKpNkNm3aNK29mJIkSZJUFVSZJFOSJEmSVP5MMiVJkiRJKWOSKUmSJElKGZNMSZIkSVLKpO05mYV/vAZWf5e6CmvWIfPy/0tdfcCQIUM47rjj6Ny581b3+9e//sXVV1/Nu+++y5gxY+jWrVtK45AkSZKkyiJtSWZKE8zyqA8YOnRoidsLCwvJzMzctL7//vvz+OOPc88996Q8BkmSJEmqTKrMcNm8vDxatGhBz549yc7Oplu3bqxevRpIJJPt2rWjVatW9O3blxgjAL1792bcuHFA4hEogwcPpsoieXsAABpNSURBVG3btowdO3azups2bcqPf/xjMjKqzMcpSZIkSSWqUlnRggUL6N+/P/Pnz6dOnTqMGDECgAEDBjBz5kxyc3NZs2YNkyZNKvH4+vXrM3v2bHr06LEzw5YkSZKkSqNKJZlNmjShQ4cOAPTq1Yvp06cDMHXqVNq3b0/r1q2ZMmUK8+bNK/H47t2777RYJUmSJKkySt+czDQIIWyxXlBQQP/+/Zk1axZNmjTh1ltvpaCgoMTja9WqtTPClCRJkqRKq0r1ZC5cuJAZM2YAMHr0aDp27LgpoWzQoAH5+fmb5mBKkiRJkrZf+pLMmnV2en3Nmzdn+PDhZGdns3z5cvr168cee+xBnz59aNWqFaeccgrt2rXb7lPPnDmT/fbbj7Fjx3LZZZdx2GGH/ZB3IEmSJEmVXth4J9UU26LS+fPnk52dXR7nKpO8vDxycnLIzc1NWwzp/gwkSZIkKUVCaQVVarisJEmSJKl8VZkks2nTpmntxZQkSZKkqqDKJJmSJEmSpPJnkilJkiRJShmTTEmSJElSyphkSpIkSZJSJitdJ/7jM++wumB9yuqrWSOLy7u3SVl9AEOGDOG4446jc+fOW93vvvvu45FHHiErK4uGDRvy2GOPccABB6Q0FkmSJEmqDNLWk5nKBLM86gMYOnRoiQlmYWHhZutHHHEEs2bN4t1336Vbt25cd911KY9FkiRJkiqDKjNcNi8vjxYtWtCzZ0+ys7Pp1q0bq1evBhLJZLt27WjVqhV9+/YlxghA7969GTduHJB4BMrgwYNp27YtY8eO3azuE044gZo1awJw9NFHs3jx4p34ziRJkiSp4qgySSbAggUL6N+/P/Pnz6dOnTqMGDECgAEDBjBz5kxyc3NZs2YNkyZNKvH4+vXrM3v2bHr06FHqOR599FFOPfXUcolfkiRJkiq6KpVkNmnShA4dOgDQq1cvpk+fDsDUqVNp3749rVu3ZsqUKcybN6/E47t3777V+p988klmzZrFoEGDUhu4JEmSJFUSabvxTzqEELZYLygooH///syaNYsmTZpw6623UlBQUOLxtWrVKrXul19+mdtvv53XXnuN6tWrpzRuSZIkSaosqlRP5sKFC5kxYwYAo0ePpmPHjpsSygYNGpCfn79pDub2ePvtt7nsssuYOHEijRo1SmnMkiRJklSZpC3JrFkjtZ2oZamvefPmDB8+nOzsbJYvX06/fv3YY4896NOnD61ateKUU06hXbt2233uQYMGkZ+fzznnnEObNm04/fTTf8hbkCRJkqRKL2y8k2qKbVHp/Pnzyc7OLo9zlUleXh45OTnk5uamLYZ0fwaSJEmSlCKhtIIqNVxWkiRJklS+qkyS2bRp07T2YkqSJElSVVBlkkxJkiRJUvkzyZQkSZIkpYxJpiRJkiQpZUwyJUmSJEkpk9qHVW6H/w5/gw2r16Wsvoya1Wh0RfuU1QcwZMgQjjvuODp37rzV/f74xz8yfPhwMjMzqV27Ng8//DAtW7ZMaSySJEmSVBmk7TmZX/5+espP2nhQx5TXWZLCwkIyMzM3rX/33XfUqVMHgIkTJzJixAhefPHFLY7zOZmSJEmSdhE+JzMvL48WLVrQs2dPsrOz6datG6tXrwZg6NChtGvXjlatWtG3b182Jt69e/dm3LhxQOIRKIMHD6Zt27aMHTt2s7o3JpgAq1atIoRSP29JkiRJ2qVVmSQTYMGCBfTv35/58+dTp04dRowYAcCAAQOYOXMmubm5rFmzhkmTJpV4fP369Zk9ezY9evTYomz48OEcdNBBXHfddfzhD38o1/chSZIkSRVVlUoymzRpQocOHQDo1asX06cnhuxOnTqV9u3b07p1a6ZMmcK8efNKPL579+6l1n3FFVfw8ccfc9ddd/G73/0u9cFLkiRJUiVQpZLM4sNYQwgUFBTQv39/xo0bx9y5c+nTpw8FBQUlHl+rVq1tnqNHjx688MILKYlXkiRJkiqbKpVkLly4kBkzZgAwevRoOnbsuCmhbNCgAfn5+ZvmYG6PDz/8cNPryZMnc8ghh6QmYEmSJEmqZNL2CJOMmtVS/giTbWnevDnDhw/n4osvpmXLlvTr14+aNWvSp08fWrVqRePGjWnXrt12n/uhhx7i5Zdfplq1atSrV48nnnjih7wFSZIkSar00vYIk50tLy+PnJwccnNz0xZDuj8DSZIkSUoRH2EiSZIkSSp/VSbJbNq0aVp7MSVJkiSpKqgySaYkSZIkqfyZZEqSJEmSUsYkU5IkSZKUMiaZkiRJkqSUSdtzMp//6FIKClekrL4amXU56+BHUlYfwJAhQzjuuOPo3LlzmfYfP3483bp1Y+bMmRx55JEpjUWSJEmSKoO0JZmpTDDLoz6AoUOHlri9sLCQzMzMzbatXLmSBx54gPbt26c8DkmSJEmqLKrMcNm8vDxatGhBz549yc7Oplu3bqxevRpIJJPt2rWjVatW9O3blxgjAL1792bcuHFA4hEogwcPpm3btowdO3aL+m+++WYGDx5MjRo1dt6bkiRJkqQKpsokmQALFiygf//+zJ8/nzp16jBixAgABgwYwMyZM8nNzWXNmjVMmjSpxOPr16/P7Nmz6dGjx2bbZ8+ezaJFizjttNPK/T1IkiRJUkVWpZLMJk2a0KFDBwB69erF9OnTAZg6dSrt27endevWTJkyhXnz5pV4fPfu3bfYtmHDBn79619z7733ll/gkiRJklRJpG1OZjqEELZYLygooH///syaNYsmTZpw6623UlBQUOLxtWrV2mLbypUryc3NpVOnTgB8+eWXnH766UycONGb/0iSJEmqcqpUT+bChQuZMWMGAKNHj6Zjx46bEsoGDRqQn5+/aQ5mWdWtW5elS5eSl5dHXl4eRx99tAmmJEmSpCorbUlmjcy6O72+5s2bM3z4cLKzs1m+fDn9+vVjjz32oE+fPrRq1YpTTjmFdu3apTQuSZIkSapKwsY7qabYFpXOnz+f7Ozs8jhXmeTl5ZGTk0Nubm7aYkj3ZyBJkiRJKRJKK6hSw2UlSZIkSeWryiSZTZs2TWsvpiRJkiRVBVUmyZQkSZIklT+TTEmSJElSyphkSpIkSZJSxiRTkiRJkpQyWWk788cnQeGy1NWXWR8OeiV19QFDhgzhuOOOo3Pnzlvd7/HHH2fQoEHsu+++AAwYMIBLL700pbFIkiRJUmWQviQzlQlmedQHDB06tORTFRaSmZm52bbu3bvz0EMPpTwGSZIkSapMqsxw2by8PFq0aEHPnj3Jzs6mW7durF69Gkgkk+3ataNVq1b07duXGCMAvXv3Zty4cUDiESiDBw+mbdu2jB07Nm3vQ5IkSZIqsiqTZAIsWLCA/v37M3/+fOrUqcOIESOAxPDWmTNnkpuby5o1a5g0aVKJx9evX5/Zs2fTo0ePLcrGjx/Pj3/8Y7p168aiRYvK9X1IkiRJUkVVpZLMJk2a0KFDBwB69erF9OnTAZg6dSrt27endevWTJkyhXnz5pV4fPfu3Uvc3qVLF/Ly8nj33Xc5+eSTufDCC8vnDUiSJElSBVelkswQwhbrBQUF9O/fn3HjxjF37lz69OlDQUFBicfXqlWrxO3169enevXqAFx66aW89dZbqQ1ckiRJkiqJKpVkLly4kBkzZgAwevRoOnbsuCmhbNCgAfn5+ZvmYG6PL774YtPriRMnkp2dnZqAJUmSJKmSSd/dZTPrp/4RJtvQvHlzhg8fzsUXX0zLli3p168fNWvWpE+fPrRq1YrGjRvTrl277T71H/7wByZOnEhWVhZ77rknjz/++A94A5IkSZJU+YWNd1JNsS0qnT9/flp7+PLy8sjJySE3NzdtMaT7M5AkSZKkFAmlFVSp4bKSJEmSpPJVZZLMpk2bprUXU5IkSZKqgp2aZJbT0NxKoSq/d0mSJElVx05LMmvUqMGyZcuqZLIVY2TZsmXUqFEj3aFIkiRJUrnaaTf+WbduHYsXLy71GZS7uho1arDffvtRrVq1dIciSZIkSTuq1Bv/7LQkU5IkSZK0y/DuspIkSZKk8meSKUmSJElKGZNMSZIkSVLKlMuczBDCi0CDlFesjRoAS9MdhJRke1RFY5tURWJ7VEVjm1SqLI0x/rykgvK68Y/KUQhhVozxyHTHIYHtURWPbVIVie1RFY1tUjuDw2UlSZIkSSljkilJkiRJShmTzMrp4XQHIBVhe1RFY5tURWJ7VEVjm1S5c06mJEmSJCll7MmUJEmSJKWMSaYkSZIkKWVMMiVJkiRJKWOSmUIhhLtCCPNCCN+FEJaEEP4cQthzG8cMDCF8HEJYGUL4MITQv1h5DCGsDiHkF1nqllBPrWQ960vY/nAI4csQwooQwhshhBOKlHdKnqNo/f/e0c9CFUMFbZM1Qwj3hxAWJ88xOYSwf7F9fp6Me00IITeE8LMd+RxUMVTG9ug1cteVjvYYQng1hLC2WHlOsToGhRA+DyGsCiG8HEJoVqzc6+MuqjK2Sa+RKlWM0SVFCzAMOAKoBjQE/g5M3Mr+pwOrgKOT6z9Nrp9cZJ8IdCzDuR8EXgbWF9t+H/AusC+JLxWuBFYC9ZLlnYof47LrLBW0TQ4HXgcaATWBJ5JtNCNZ3gxYDfQCfgT0TMbQNN2fp0uVbI9eI3fRJR3tEXgVuGkr5T2B/wJtk+3xD0AukJks9/q4Cy+VtE16jXQpcbEnM4VijDfGGN+OMa6LMX4NPEDih680BwNzYoz/SR4/g8QfN4dvz3lDCMcBxwJ3lXKOSTHGz2OMG4A/A7WBg7bnHKqcKmibPAe4O8b43xjjamAI0BromCy/EHgrxvhkjPH7GONTwOzkdlVilbQ9aheVrva4DX2BP8UYZyfb440kEkuvj1VAJW2TUolMMsvXScCcrZSPAeqEEDqEEDJCCMcChwIvFttvbAhhaUgMdT27aEEIoSaJxPFSYF0J5/gDcFIIYf8QQhZwOfARiW+hNsoMISwKiSG1k0MIqbw4qWKpCG0yJJeNNl6H2iT/PRx4q9gxs0ntL01VDJWhPYLXyKqi3Ntj0tUhhG+SwyJvCCFUK1K22fUvxpgPfMj/rn9eH6uWytAmwWukSpCV7gB2VSGEriQSuuO3stt/gXHAVP73h83VMcaiCWBnEkO5AM4AngohnBVj3HgBuQP4a4xxVgihUwnnmAPkAZ8BhcC3wBkxxoJk+fsk/piaR6KHczAwJYTQOsa4pIxvV5VABWqTk4DBIYQ3gHzgtySG89RJlu8OrCh2zLfAYVt/h6pMKlF79BpZBezE9ngDiTb1HdAOeIpEW7shWV7a9c/rYxVTidqk10iVLN3jdXfFhcTwq2+AE7ax323AB0BLEt+kHwZ8DFyylWP+DIxKvu6YPL5mcr0TW843eg14nsR8oyygC8lfSFs5x4dbi8Gl8i0VrE3WBR4GFgGfA9eQ+OV2ebL8BeD+Ysc8ADyX7s/Rpeq1x1LO4TVyF1p2VnsspbwnsKjI+rfAmcX2mQP8Kvna62MVWCpTmyylDq+RLg6XTbUQwkXAn4AuMcap29j9J8DzMcb3YsI8Er9AumzlmA38b2hXZ6AJsDCEsBSYQGLIwtIQwsY6fgI8HBPzjdbHGP9K4gJ0chnPoUquorXJGOOKGGPfGGOTGOO+wD9JfFP6arKOOSRuMFDUEWx9yJAqiUrYHrd1DlViO7k9lqV8s+tfCKE2cAj/u/55fdzFVcI2+UPOoaog3VnurrQAvwKWAe3KuP8NwALgkOR6NokE8ObkeivgKBJ3kKsGnEnirnKnJ8vrAPsVWc4B1idf75bc5x/AeGBPEkMpTgMKgOOS5SeSmDieQWKYw60kvrVqku7P02WXbZMHAnuT+AXUAngD+EuRGA5K1vnL5Dl+iXdP3CWWStoevUbuoksa2uMeQE6yHQUSyeEC4N4i5+gJfJUs2w24n8QwxI138vT6uAsvlbRNeo10Kbl9pjuAXWkhMY9nHYl5PZuWIuU9i61nAXeSmDOZDywE7gWqJctPSP4grwKWA7OAHls5fye2HAq2D/B08gLxXbK+S4uUX0NivuYqEmP7Xyzrxc2l4i8VtE3mJNvc6mT9vwWyiu3z8+R51iT//Vm6P0uXqtkevUbuusvObo8kHknxHxLz21aSGOZ4C/CjYnFdByxJtslXgIOKlXt93EWXytgmvUa6lLaEGCOSJEmSJKWCczIlSZIkSSljkilJkiRJShmTTEmSJElSyphkSpIkSZJSxiRTkiRJkpQyJpmSJEmSVEWEEB4LIfw3hJBbhn3/L4TwTnL5IITwbVnOYZIpSSpRCCGGELqlOw5VHCGE3iGE/DLslxdCGLgzYpIkbbfHSTxzd5tijNfEGNvEGNsADwLPleU4k0xJqoJCCHuFEB4IIXwcQlgbQvg8hPD3EMIv0h1bVRBC6BRCmBBC+CKEsDqE8G4I4eJi+5QpodvJngGabVwJIdxayjfh7YAROy0qSVKZxRj/BXxTdFsI4aAQwoshhLdCCNNCCC1KOPSXwNNlOUdWCuKUJFUiIYSmwOvASuAGYA6JLx1PAv4I7L8DdVeLMa7b8Sh3eccAc4G7gS+AU4CHQwgFMcbRaY1sK2KMa4A1Zdjv650QjiQpdR4GLo8xfhhCaE/ii8ITNxaGEA4ADgSmlKUyezIlqerZ2MN0ZIzx2Rjjghjj/BjjQ8CPi+27ZwhhbAhhVQjhkxBCr40FIYSmySG1vwwhTAkhrAEuCyFkhBBuDiEsSvaSzg0hnFHCcT1CCK+FENaEEN4OIfw4hNAqhPDv5PmmhxAOLBpMCOGyEMJHIYTvk//2KaH8gxBCQQhhaQjhHyGErGRZecbVJfntb0EI4dMQwu0hhB+V9h8QYxwWY7wpxvh6jPGTGONIEkOQupblP7Akyd7RGELISc6dKUjG9JNi+52dfO9rk5/Fb0IIoVj5u8n3/03ys9grWbapdzWE0Bu4BTgsed6Y3LbFcNkQwv4hhOdDCCuTy3MhhP2KlN8aQshNfvYfJ/d5IYTQ4Id+HpKksgkh1Cbx5efYEMI7wJ+AvYvt1gMYF2MsLEudJpmSVIWEEPYkMQ9jeIxxi6GYMcbiE/qHABOAw0kMlXwshFC8p/MOEolrS+AF4CpgEDAYaA08DzwXQmhT7LjbgLuAI4BvSQzBeRD4DXAUUAP4Q5HYzwIeAu4HWgEPACNCCF2S5UcCw5P1NifRM/tikfOVV1ynAE8lYzsMuBjoBgxj+9QBlm/nMSW5h8R7PBL4BJgUQqiZjPUnwFgSCW1r4HoSvdkDkuWNgTHAE0A2cBwwqpTzPAPcCywg8cfI3sltmwkhZJBoQ3sBJySXfYAXiia3QFOgO3AW8DMSn//t2//2JUnbKQP4duPcy+SSXWyfHpRxqCwAMUYXFxcXlyqykEiSInBWGfaNwB1F1rOA1UCv5HrT5D7XFjvuc2BIsW2vAk8WO+6yIuU5yW1nF9nWG8gvsv468Fixeh8Hpidfnw2sAHYv5f2UV1z/Am4uVu+ZQD4Qyvj/kgOsA44q7TxlqKNTMtaeRbbVJpEoX5pcfwqYUuy4W4HFyddtk3UcUMo5ir/3W4HcEvbLAwYmX58MFAJNi5Q3AzYAnYvUUwDULbLPb4CP0v0z4+Li4rIrLsnfeblF1v8NnJN8HYDDi5S1SF7Xy/Q7LcZoT6YkVTFh27ts5t2NL2KM64GvgUbF9pm1qfIQ6pDopXq92D7TSfR0llg38FXy37nFttXa2AtHomdta/W+BHwGfBpCeCqEcGEIYfedENdPgN+EEPI3LsBooBbQmG0IIXRI7v+rGOOb29q/DGZsfBETvdVz+d97LO0z3Df5Gc0BXgZyQwjjQwj9QggNdzCebGBJjDGvSFyfAEvY/LP/LMa4osj6ErZsa5KkHRRCeJrE74rmIYTFIYRLgJ7AJSGEOcA84Iwih/QAxsRkxlkW3vhHkqqWD0n0VGWTGC66LcVv4hPZcqrFqjKeu/gvp3UllJW0bVtfiCa6XWNcGUJoS2KI58kkhoEOCyG0I9GrWF5xZZAYYju2hLq3egOcEEJH4G8kelhHbm3fnSDGGAtDCD8DjiYxZPUS4I4QwvExxjnlcc4ir8vS1iRJOyjG+MtSikp8rEmM8dbtPYcXb0mqQmKM3wD/AAYkJ/pvJoSwxw7W/x2JHqgOxYo6Au/tSN3A/G3VG2NcH2OcEmO8gcRNjGoBOeUc12ygRYzxoxKW9aUdFEI4Dvg7cGuM8f4djKGoo4ucoxaJ+avzk5tK+wwXxxhXQiLTjDHOiDHeRuJRJEtIzJUsyfdA5jbimQ/sExJ3Nd4YVzMSPcs7+tlLkiogezIlqeq5gsSQyVkhhJtJDA8NJG7IcgM78AiTpN8DQ0MIHwJvAb2AY0nM99vReseGEN4C/kniG9eeJOZiEkLIAQ4iMUfyGxLvZ3f+l2CVV1xDSdxc5zPgWWA9icTuqBjjdSUdEELoBEwmccOk0ckb7gAUxs0f/5FRwo2J1scYS3o25UY3hRC+JpEcDiGRCG58LMq9wMwQwq3Jbe2Aa4Ebk3EdDXQm8UXEVyRuvtOE0pPBPOCAZA/yQmBljHFtsX1eJtHGngohXJXc9iCJ5LxMt8KXJFUuJpmSVMXEGD9JJgU3kriL6r7AMhLz8fqm4BR/IJHc3U3ijqILgK47OtwyxvhCCOFKYCCJO8x+BvSPMf41ucu3JG64MwSoCXxM4oY308o5rn+EEE4Dbk7Gth74gMRNiUrTOxnjwOSy0Wckbsaw0W7A28WOXQZs7dEe15NIJpuTmFeTE2NclYx1dgjhHBLDe28kkUjeSeLOuJC4cVIH4EpgD2AR8NsY45OlnGs8iST/leT+F1HsfccYY/JRMX8ApiY3vwxcuT3zeyRJlUfw+i5JUuWX7B2dCjSMMS5NcziSpCrMOZmSJEmSpJQxyZQkSZIkpYzDZSVJkiRJKWNPpiRJkiQpZUwyJUmSJEkpY5IpSZIkSUoZk0xJkiRJUsqYZEqSJEmSUub/A/z54KkuqGFGAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1152x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "AgamPrimer.plot_primer_locs(primer_res_dict=results_dict, assay_type=assay_type, gff=ag3.geneset(), contig=contig, seq_parameters=seq_parameters, n_primer_pairs=n_primer_pairs, legend_loc='lower left')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0JA92ApBcxLo",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 328
    },
    "id": "0JA92ApBcxLo",
    "outputId": "72f39feb-d67e-4f53-d161-303bc939a4f8"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-16a95349-f358-4e3e-9663-2fc8b990a3e6\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>primer_pair</th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>parameter</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>PRIMER_LEFT_SEQUENCE</th>\n",
       "      <td>GACATCGCTCTACGGGACTG</td>\n",
       "      <td>CGTTTTAAGGATCCCACGCC</td>\n",
       "      <td>CTGGGTACGTCAGCTTCCAG</td>\n",
       "      <td>CTTCGCCGGAAAGCATTGTG</td>\n",
       "      <td>GCCGGAAAGCATTGTGGATC</td>\n",
       "      <td>TTTTAAGGATCCCACGCCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_RIGHT_SEQUENCE</th>\n",
       "      <td>GGTGGCGTGGGATCCTTAAA</td>\n",
       "      <td>GGGTCACATTGCTCGGTACA</td>\n",
       "      <td>GGCGTGGGATCCTTAAAACG</td>\n",
       "      <td>CTTGTATTTCCTGCGGTCCG</td>\n",
       "      <td>GCCTTCTTGTATTTCCTGCGG</td>\n",
       "      <td>ACGCGCCGATCAAAATGAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_LEFT_TM</th>\n",
       "      <td>59.970433</td>\n",
       "      <td>59.26916</td>\n",
       "      <td>60.108615</td>\n",
       "      <td>60.454957</td>\n",
       "      <td>59.898781</td>\n",
       "      <td>60.61485</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_RIGHT_TM</th>\n",
       "      <td>60.034863</td>\n",
       "      <td>60.036765</td>\n",
       "      <td>59.26916</td>\n",
       "      <td>58.993168</td>\n",
       "      <td>59.598095</td>\n",
       "      <td>59.480481</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_LEFT_GC_PERCENT</th>\n",
       "      <td>60.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>52.380952</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_RIGHT_GC_PERCENT</th>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>52.380952</td>\n",
       "      <td>45.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_LEFT</th>\n",
       "      <td>(249, 20)</td>\n",
       "      <td>(322, 20)</td>\n",
       "      <td>(266, 20)</td>\n",
       "      <td>(962, 20)</td>\n",
       "      <td>(966, 20)</td>\n",
       "      <td>(324, 21)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_RIGHT</th>\n",
       "      <td>(344, 20)</td>\n",
       "      <td>(389, 20)</td>\n",
       "      <td>(341, 20)</td>\n",
       "      <td>(1021, 20)</td>\n",
       "      <td>(1026, 21)</td>\n",
       "      <td>(413, 20)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PRIMER_PAIR_PRODUCT_SIZE</th>\n",
       "      <td>96</td>\n",
       "      <td>68</td>\n",
       "      <td>76</td>\n",
       "      <td>60</td>\n",
       "      <td>61</td>\n",
       "      <td>90</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-16a95349-f358-4e3e-9663-2fc8b990a3e6')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-16a95349-f358-4e3e-9663-2fc8b990a3e6 button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-16a95349-f358-4e3e-9663-2fc8b990a3e6');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "primer_pair                                  0                     1  \\\n",
       "parameter                                                              \n",
       "PRIMER_LEFT_SEQUENCE      GACATCGCTCTACGGGACTG  CGTTTTAAGGATCCCACGCC   \n",
       "PRIMER_RIGHT_SEQUENCE     GGTGGCGTGGGATCCTTAAA  GGGTCACATTGCTCGGTACA   \n",
       "PRIMER_LEFT_TM                       59.970433              59.26916   \n",
       "PRIMER_RIGHT_TM                      60.034863             60.036765   \n",
       "PRIMER_LEFT_GC_PERCENT                    60.0                  55.0   \n",
       "PRIMER_RIGHT_GC_PERCENT                   55.0                  55.0   \n",
       "PRIMER_LEFT                          (249, 20)             (322, 20)   \n",
       "PRIMER_RIGHT                         (344, 20)             (389, 20)   \n",
       "PRIMER_PAIR_PRODUCT_SIZE                    96                    68   \n",
       "\n",
       "primer_pair                                  2                     3  \\\n",
       "parameter                                                              \n",
       "PRIMER_LEFT_SEQUENCE      CTGGGTACGTCAGCTTCCAG  CTTCGCCGGAAAGCATTGTG   \n",
       "PRIMER_RIGHT_SEQUENCE     GGCGTGGGATCCTTAAAACG  CTTGTATTTCCTGCGGTCCG   \n",
       "PRIMER_LEFT_TM                       60.108615             60.454957   \n",
       "PRIMER_RIGHT_TM                       59.26916             58.993168   \n",
       "PRIMER_LEFT_GC_PERCENT                    60.0                  55.0   \n",
       "PRIMER_RIGHT_GC_PERCENT                   55.0                  55.0   \n",
       "PRIMER_LEFT                          (266, 20)             (962, 20)   \n",
       "PRIMER_RIGHT                         (341, 20)            (1021, 20)   \n",
       "PRIMER_PAIR_PRODUCT_SIZE                    76                    60   \n",
       "\n",
       "primer_pair                                   4                      5  \n",
       "parameter                                                               \n",
       "PRIMER_LEFT_SEQUENCE       GCCGGAAAGCATTGTGGATC  TTTTAAGGATCCCACGCCACC  \n",
       "PRIMER_RIGHT_SEQUENCE     GCCTTCTTGTATTTCCTGCGG   ACGCGCCGATCAAAATGAAA  \n",
       "PRIMER_LEFT_TM                        59.898781               60.61485  \n",
       "PRIMER_RIGHT_TM                       59.598095              59.480481  \n",
       "PRIMER_LEFT_GC_PERCENT                     55.0              52.380952  \n",
       "PRIMER_RIGHT_GC_PERCENT               52.380952                   45.0  \n",
       "PRIMER_LEFT                           (966, 20)              (324, 21)  \n",
       "PRIMER_RIGHT                         (1026, 21)              (413, 20)  \n",
       "PRIMER_PAIR_PRODUCT_SIZE                     61                     90  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primer_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "-aIy4IyoBc7-",
   "metadata": {
    "id": "-aIy4IyoBc7-"
   },
   "source": [
    "###**Checking our primers for specificity to the *Anopheles gambiae* genome**\n",
    "\n",
    "We can use a cool new python package, gget, to rapidly search our primers against the AgamP3 genome, to ensure they will not amplify any other regions of the genome. gget rapidly queries large databases - in this case we can use the program BLAT to align primer sequences."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "Fub-U39VTpxj",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 34
    },
    "id": "Fub-U39VTpxj",
    "outputId": "a8e68306-a8c2-470e-e5e8-31942d51f675"
   },
   "outputs": [
    {
     "data": {
      "application/vnd.google.colaboratory.intrinsic+json": {
       "type": "string"
      },
      "text/plain": [
       "'GACATCGCTCTACGGGACTG'"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "primer_sequence = primer_df.loc['PRIMER_LEFT_SEQUENCE', str(0)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "v9ExsM7cBcDm",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/",
     "height": 74
    },
    "id": "v9ExsM7cBcDm",
    "outputId": "b0eaf6de-b568-4cd5-a120-4abc23fe2a48"
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "  <div id=\"df-18fc4e25-322b-4797-bf71-f796f821e63e\">\n",
       "    <div class=\"colab-df-container\">\n",
       "      <div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genome</th>\n",
       "      <th>query_size</th>\n",
       "      <th>aligned_start</th>\n",
       "      <th>aligned_end</th>\n",
       "      <th>matches</th>\n",
       "      <th>mismatches</th>\n",
       "      <th>%_aligned</th>\n",
       "      <th>%_matched</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>strand</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>anoGam3</td>\n",
       "      <td>22</td>\n",
       "      <td>1</td>\n",
       "      <td>22</td>\n",
       "      <td>22</td>\n",
       "      <td>0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>100.0</td>\n",
       "      <td>chr2L</td>\n",
       "      <td>+</td>\n",
       "      <td>2422598</td>\n",
       "      <td>2422619</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>\n",
       "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-18fc4e25-322b-4797-bf71-f796f821e63e')\"\n",
       "              title=\"Convert this dataframe to an interactive table.\"\n",
       "              style=\"display:none;\">\n",
       "        \n",
       "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
       "       width=\"24px\">\n",
       "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
       "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
       "  </svg>\n",
       "      </button>\n",
       "      \n",
       "  <style>\n",
       "    .colab-df-container {\n",
       "      display:flex;\n",
       "      flex-wrap:wrap;\n",
       "      gap: 12px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert {\n",
       "      background-color: #E8F0FE;\n",
       "      border: none;\n",
       "      border-radius: 50%;\n",
       "      cursor: pointer;\n",
       "      display: none;\n",
       "      fill: #1967D2;\n",
       "      height: 32px;\n",
       "      padding: 0 0 0 0;\n",
       "      width: 32px;\n",
       "    }\n",
       "\n",
       "    .colab-df-convert:hover {\n",
       "      background-color: #E2EBFA;\n",
       "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
       "      fill: #174EA6;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert {\n",
       "      background-color: #3B4455;\n",
       "      fill: #D2E3FC;\n",
       "    }\n",
       "\n",
       "    [theme=dark] .colab-df-convert:hover {\n",
       "      background-color: #434B5C;\n",
       "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
       "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
       "      fill: #FFFFFF;\n",
       "    }\n",
       "  </style>\n",
       "\n",
       "      <script>\n",
       "        const buttonEl =\n",
       "          document.querySelector('#df-18fc4e25-322b-4797-bf71-f796f821e63e button.colab-df-convert');\n",
       "        buttonEl.style.display =\n",
       "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
       "\n",
       "        async function convertToInteractive(key) {\n",
       "          const element = document.querySelector('#df-18fc4e25-322b-4797-bf71-f796f821e63e');\n",
       "          const dataTable =\n",
       "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
       "                                                     [key], {});\n",
       "          if (!dataTable) return;\n",
       "\n",
       "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
       "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
       "            + ' to learn more about interactive tables.';\n",
       "          element.innerHTML = '';\n",
       "          dataTable['output_type'] = 'display_data';\n",
       "          await google.colab.output.renderOutput(dataTable, element);\n",
       "          const docLink = document.createElement('div');\n",
       "          docLink.innerHTML = docLinkHtml;\n",
       "          element.appendChild(docLink);\n",
       "        }\n",
       "      </script>\n",
       "    </div>\n",
       "  </div>\n",
       "  "
      ],
      "text/plain": [
       "    genome  query_size  aligned_start  aligned_end  matches  mismatches  \\\n",
       "0  anoGam3          22              1           22       22           0   \n",
       "\n",
       "   %_aligned  %_matched chromosome strand    start      end  \n",
       "0      100.0      100.0      chr2L      +  2422598  2422619  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gget.blat(sequence=primer_sequence, seqtype='DNA', assembly='anoGam3')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "UzoL2FPQCdSg",
   "metadata": {
    "id": "UzoL2FPQCdSg"
   },
   "source": [
    "We can even blast the sequence against full blast databases, although so far I've struggled to get hits with the default settings (though I havent had a change to play around much!) "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "zP4LB2bQBgGg",
   "metadata": {
    "colab": {
     "base_uri": "https://localhost:8080/"
    },
    "id": "zP4LB2bQBgGg",
    "outputId": "0f32aa8f-023f-4e4f-be04-4f4030649f66"
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:No significant similarity found for search GEJ7880J016. If your sequence is very short, try increasing the 'expect' argument.\n"
     ]
    }
   ],
   "source": [
    "gget.blast(sequence='CGGTGATGTATCCTGCATACCA', program='blastn', database='nt', expect=200)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "g4r9sQectXxx",
   "metadata": {
    "id": "g4r9sQectXxx"
   },
   "source": [
    "####**We may now have designed suitable primers. However, there are some further considerations...**\n",
    "\n",
    "\n",
    "- Primers **must** be run in [**NCBI Blast**](https://blast.ncbi.nlm.nih.gov/Blast.cgi), to ensure specificity against the host organism, and specificity for the genomic location of interest. I would recommend both doing a general blast and also specifically against *An. gambiae* (TaxonID = 7165).\n",
    "\n",
    "- If in multiplexed use with other primers or probes, primers must not interact with each other. This can be investigated on a one by one basis using the IDT tool [oligoanalyzer](https://eu.idtdna.com/calc/analyzer), though higher throughput algorithms may be required.\n",
    "\n",
    "- If designing Locked Nucleic Acid (LNA) probes for SNP detection, you will want to play around with the placement of LNAs in the oligo sequence, which can allow short probes (~10-14 bases) to bind with high affinity and discriminate between SNPs. IDT have a tool for this which allow you to check the binding affinity between mismatches, though it requires a log in https://eu.idtdna.com/calc/analyzer/lna. \n",
    "\n",
    "- Many more considerations.... [IDT - How to design primers and probes for PCR and qPCR](https://eu.idtdna.com/pages/education/decoded/article/designing-pcr-primers-and-probes)   \n",
    "\n",
    "\\\n",
    "  \n",
    "\n",
    "####**Future development**\n",
    "\n",
    "Any contributions or suggestions on how we can improve this notebook are more than welcome. Please [email](mailto:sanjay.nagi@lstmed.ac.uk) or log an [issue on github](https://github.com/sanjaynagi/primerDesignAg/issues). This notebook and source code for AgamPrimer are located here - https://github.com/sanjaynagi/AgamPrimer/    \n",
    "\n",
    "\\\n",
    "####**References**\n",
    "\n",
    "The Anopheles gambiae 1000 Genomes Consortium (2020). **Genome variation and population structure among 1142 mosquitoes of the African malaria vector species *Anopheles gambiae* and *Anopheles coluzzii***. *Genome Research*, **30**: 1533-1546. \n",
    "https://genome.cshlp.org/content/early/2020/09/25/gr.262790.120\n",
    "\n",
    "Untergasser A, Cutcutache I, Koressaar T, Ye J, Faircloth BC, Remm M and Rozen SG (2012). **Primer3--new capabilities and interfaces**. *Nucleic Acids Research*. 40(15):e115."
   ]
  }
 ],
 "metadata": {
  "colab": {
   "include_colab_link": true,
   "name": "Primer Design in Anopheles gambiae.ipynb",
   "provenance": []
  },
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.12"
  },
  "vscode": {
   "interpreter": {
    "hash": "710100144729ec1057471750b60f3250df4296be7eb29e7e87fec32ed718ec4b"
   }
  },
  "widgets": {
   "application/vnd.jupyter.widget-state+json": {
    "2db4d9dd91a24e32b80d870e9869339a": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "4546f79977e24400806287bebf4efca0": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "506dbb25d1f743ee80033d6f98dabc3b": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_2db4d9dd91a24e32b80d870e9869339a",
      "placeholder": "",
      "style": "IPY_MODEL_cbf3c833b87d4f51b10aa6f09d53d5d1",
      "value": "Load sample metadata: 100%"
     }
    },
    "5a98a29b6ecf41f09fee8d7aebf3132e": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "FloatProgressModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "FloatProgressModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "ProgressView",
      "bar_style": "success",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_a60370f9435b42ab8b5d794ce32c2fa2",
      "max": 28,
      "min": 0,
      "orientation": "horizontal",
      "style": "IPY_MODEL_cc3c6714c3ab435ca820f8081878dc3f",
      "value": 28
     }
    },
    "743f1f0ae54f4f4aa4994c96a523625c": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HTMLModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HTMLModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HTMLView",
      "description": "",
      "description_tooltip": null,
      "layout": "IPY_MODEL_4546f79977e24400806287bebf4efca0",
      "placeholder": "",
      "style": "IPY_MODEL_b913e8c4b6ca4c129949878e1b832774",
      "value": " 28/28 [00:43&lt;00:00,  1.37s/it]"
     }
    },
    "792f088bb26346baa2f17f4cfcafe0ea": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "a60370f9435b42ab8b5d794ce32c2fa2": {
     "model_module": "@jupyter-widgets/base",
     "model_module_version": "1.2.0",
     "model_name": "LayoutModel",
     "state": {
      "_model_module": "@jupyter-widgets/base",
      "_model_module_version": "1.2.0",
      "_model_name": "LayoutModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "LayoutView",
      "align_content": null,
      "align_items": null,
      "align_self": null,
      "border": null,
      "bottom": null,
      "display": null,
      "flex": null,
      "flex_flow": null,
      "grid_area": null,
      "grid_auto_columns": null,
      "grid_auto_flow": null,
      "grid_auto_rows": null,
      "grid_column": null,
      "grid_gap": null,
      "grid_row": null,
      "grid_template_areas": null,
      "grid_template_columns": null,
      "grid_template_rows": null,
      "height": null,
      "justify_content": null,
      "justify_items": null,
      "left": null,
      "margin": null,
      "max_height": null,
      "max_width": null,
      "min_height": null,
      "min_width": null,
      "object_fit": null,
      "object_position": null,
      "order": null,
      "overflow": null,
      "overflow_x": null,
      "overflow_y": null,
      "padding": null,
      "right": null,
      "top": null,
      "visibility": null,
      "width": null
     }
    },
    "b913e8c4b6ca4c129949878e1b832774": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "cbf3c833b87d4f51b10aa6f09d53d5d1": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "DescriptionStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "DescriptionStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "description_width": ""
     }
    },
    "cc3c6714c3ab435ca820f8081878dc3f": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "ProgressStyleModel",
     "state": {
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "ProgressStyleModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/base",
      "_view_module_version": "1.2.0",
      "_view_name": "StyleView",
      "bar_color": null,
      "description_width": ""
     }
    },
    "e2fec3b726cd4f529b58b0716f2237a3": {
     "model_module": "@jupyter-widgets/controls",
     "model_module_version": "1.5.0",
     "model_name": "HBoxModel",
     "state": {
      "_dom_classes": [],
      "_model_module": "@jupyter-widgets/controls",
      "_model_module_version": "1.5.0",
      "_model_name": "HBoxModel",
      "_view_count": null,
      "_view_module": "@jupyter-widgets/controls",
      "_view_module_version": "1.5.0",
      "_view_name": "HBoxView",
      "box_style": "",
      "children": [
       "IPY_MODEL_506dbb25d1f743ee80033d6f98dabc3b",
       "IPY_MODEL_5a98a29b6ecf41f09fee8d7aebf3132e",
       "IPY_MODEL_743f1f0ae54f4f4aa4994c96a523625c"
      ],
      "layout": "IPY_MODEL_792f088bb26346baa2f17f4cfcafe0ea"
     }
    }
   }
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
